<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Data variety | Environmental Systems Data Science</title>
  <meta name="description" content="Text book and exercises for the course Environmental System Data Science at ETH Zürich." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Data variety | Environmental Systems Data Science" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Text book and exercises for the course Environmental System Data Science at ETH Zürich." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Data variety | Environmental Systems Data Science" />
  
  <meta name="twitter:description" content="Text book and exercises for the course Environmental System Data Science at ETH Zürich." />
  

<meta name="author" content="Loïc Pellissier, Joshua Payne, Benjamin Stocker" />


<meta name="date" content="2021-03-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="ch-02.html"/>
<link rel="next" href="ch-04.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Prerequisites</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-description"><i class="fa fa-check"></i>Course Description</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-objectives"><i class="fa fa-check"></i>Course Objectives</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#content"><i class="fa fa-check"></i>Content</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#useful-prerequisites"><i class="fa fa-check"></i>Useful Prerequisites</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="ch-01.html"><a href="ch-01.html"><i class="fa fa-check"></i><b>1</b> Primers</a>
<ul>
<li class="chapter" data-level="1.1" data-path="ch-01.html"><a href="ch-01.html#introduction"><i class="fa fa-check"></i><b>1.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="ch-01.html"><a href="ch-01.html#important-points-from-the-lecture"><i class="fa fa-check"></i><b>1.1.1</b> Important points from the lecture</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="ch-01.html"><a href="ch-01.html#tutorial"><i class="fa fa-check"></i><b>1.2</b> Tutorial</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="ch-01.html"><a href="ch-01.html#working-with-jupyter-notebook"><i class="fa fa-check"></i><b>1.2.1</b> Working with Jupyter notebook</a></li>
<li class="chapter" data-level="1.2.2" data-path="ch-01.html"><a href="ch-01.html#using-git-for-version-control"><i class="fa fa-check"></i><b>1.2.2</b> Using Git for version control</a></li>
<li class="chapter" data-level="1.2.3" data-path="ch-01.html"><a href="ch-01.html#basics-of-r-code"><i class="fa fa-check"></i><b>1.2.3</b> Basics of R code</a></li>
<li class="chapter" data-level="1.2.4" data-path="ch-01.html"><a href="ch-01.html#reading-data-into-r"><i class="fa fa-check"></i><b>1.2.4</b> Reading data into R</a></li>
<li class="chapter" data-level="1.2.5" data-path="ch-01.html"><a href="ch-01.html#r-objects"><i class="fa fa-check"></i><b>1.2.5</b> R Objects</a></li>
<li class="chapter" data-level="1.2.6" data-path="ch-01.html"><a href="ch-01.html#data-visualisation"><i class="fa fa-check"></i><b>1.2.6</b> Data visualisation</a></li>
<li class="chapter" data-level="1.2.7" data-path="ch-01.html"><a href="ch-01.html#example-code-loops"><i class="fa fa-check"></i><b>1.2.7</b> Example code: Loops</a></li>
<li class="chapter" data-level="1.2.8" data-path="ch-01.html"><a href="ch-01.html#where-to-find-help"><i class="fa fa-check"></i><b>1.2.8</b> Where to find Help</a></li>
<li class="chapter" data-level="1.2.9" data-path="ch-01.html"><a href="ch-01.html#further-reading"><i class="fa fa-check"></i><b>1.2.9</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="ch-01.html"><a href="ch-01.html#exercise"><i class="fa fa-check"></i><b>1.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="ch-02.html"><a href="ch-02.html"><i class="fa fa-check"></i><b>2</b> Data wrangling</a>
<ul>
<li class="chapter" data-level="2.1" data-path="ch-02.html"><a href="ch-02.html#introduction-1"><i class="fa fa-check"></i><b>2.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="ch-02.html"><a href="ch-02.html#data-transformation-with-dplyr"><i class="fa fa-check"></i><b>2.1.1</b> Data transformation with dplyr</a></li>
<li class="chapter" data-level="2.1.2" data-path="ch-02.html"><a href="ch-02.html#data-visualisation-with-ggplot2"><i class="fa fa-check"></i><b>2.1.2</b> Data visualisation with ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="ch-02.html"><a href="ch-02.html#tutorial-1"><i class="fa fa-check"></i><b>2.2</b> Tutorial</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="ch-02.html"><a href="ch-02.html#dataset-1-half-hourly-flux-data"><i class="fa fa-check"></i><b>2.2.1</b> Dataset 1 (half-hourly flux data)</a></li>
<li class="chapter" data-level="2.2.2" data-path="ch-02.html"><a href="ch-02.html#dataset-2-daily-flux-data"><i class="fa fa-check"></i><b>2.2.2</b> Dataset 2 (daily flux data)</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="ch-02.html"><a href="ch-02.html#exercise-1"><i class="fa fa-check"></i><b>2.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="ch-03.html"><a href="ch-03.html"><i class="fa fa-check"></i><b>3</b> Data variety</a>
<ul>
<li class="chapter" data-level="3.1" data-path="ch-03.html"><a href="ch-03.html#introduction-2"><i class="fa fa-check"></i><b>3.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="ch-03.html"><a href="ch-03.html#overview"><i class="fa fa-check"></i><b>3.1.1</b> Overview</a></li>
<li class="chapter" data-level="3.1.2" data-path="ch-03.html"><a href="ch-03.html#learning-objectives"><i class="fa fa-check"></i><b>3.1.2</b> Learning objectives</a></li>
<li class="chapter" data-level="3.1.3" data-path="ch-03.html"><a href="ch-03.html#key-points-of-the-lecture"><i class="fa fa-check"></i><b>3.1.3</b> Key points of the lecture</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="ch-03.html"><a href="ch-03.html#tutorial-2"><i class="fa fa-check"></i><b>3.2</b> Tutorial</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="ch-03.html"><a href="ch-03.html#overview-1"><i class="fa fa-check"></i><b>3.2.1</b> Overview</a></li>
<li class="chapter" data-level="3.2.2" data-path="ch-03.html"><a href="ch-03.html#modis-remote-download"><i class="fa fa-check"></i><b>3.2.2</b> MODIS remote download</a></li>
<li class="chapter" data-level="3.2.3" data-path="ch-03.html"><a href="ch-03.html#points-on-the-globe"><i class="fa fa-check"></i><b>3.2.3</b> Points on the globe</a></li>
<li class="chapter" data-level="3.2.4" data-path="ch-03.html"><a href="ch-03.html#shapefiles"><i class="fa fa-check"></i><b>3.2.4</b> Shapefiles</a></li>
<li class="chapter" data-level="3.2.5" data-path="ch-03.html"><a href="ch-03.html#rasters"><i class="fa fa-check"></i><b>3.2.5</b> Rasters</a></li>
<li class="chapter" data-level="3.2.6" data-path="ch-03.html"><a href="ch-03.html#key-points-of-the-tutorial"><i class="fa fa-check"></i><b>3.2.6</b> Key points of the tutorial</a></li>
<li class="chapter" data-level="3.2.7" data-path="ch-03.html"><a href="ch-03.html#bonus-species-occurrence-trait-data-and-pcas"><i class="fa fa-check"></i><b>3.2.7</b> Bonus: Species Occurrence, Trait Data and PCAs</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="ch-03.html"><a href="ch-03.html#exercise-2"><i class="fa fa-check"></i><b>3.3</b> Exercise</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="ch-03.html"><a href="ch-03.html#part-1-plotting-elevation-differences"><i class="fa fa-check"></i><b>3.3.1</b> Part 1: Plotting Elevation differences</a></li>
<li class="chapter" data-level="3.3.2" data-path="ch-03.html"><a href="ch-03.html#part-2-temperature-and-elevation-correlations"><i class="fa fa-check"></i><b>3.3.2</b> Part 2: Temperature and Elevation Correlations</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="ch-04.html"><a href="ch-04.html"><i class="fa fa-check"></i><b>4</b> Data Scraping</a>
<ul>
<li class="chapter" data-level="4.1" data-path="ch-04.html"><a href="ch-04.html#introduction-3"><i class="fa fa-check"></i><b>4.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="ch-04.html"><a href="ch-04.html#theory"><i class="fa fa-check"></i><b>4.1.1</b> Theory</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="ch-04.html"><a href="ch-04.html#tutorial-3"><i class="fa fa-check"></i><b>4.2</b> Tutorial</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="ch-04.html"><a href="ch-04.html#r-packages-and-functions"><i class="fa fa-check"></i><b>4.2.1</b> R-Packages and Functions</a></li>
<li class="chapter" data-level="4.2.2" data-path="ch-04.html"><a href="ch-04.html#the-fishbase-website"><i class="fa fa-check"></i><b>4.2.2</b> The FishBase website</a></li>
<li class="chapter" data-level="4.2.3" data-path="ch-04.html"><a href="ch-04.html#accessing-fishbase"><i class="fa fa-check"></i><b>4.2.3</b> Accessing FishBase</a></li>
<li class="chapter" data-level="4.2.4" data-path="ch-04.html"><a href="ch-04.html#scraping-numbers"><i class="fa fa-check"></i><b>4.2.4</b> Scraping Numbers</a></li>
<li class="chapter" data-level="4.2.5" data-path="ch-04.html"><a href="ch-04.html#scraping-text-snippetrs"><i class="fa fa-check"></i><b>4.2.5</b> Scraping Text Snippetrs</a></li>
<li class="chapter" data-level="4.2.6" data-path="ch-04.html"><a href="ch-04.html#scraping-tables"><i class="fa fa-check"></i><b>4.2.6</b> Scraping Tables</a></li>
<li class="chapter" data-level="4.2.7" data-path="ch-04.html"><a href="ch-04.html#the-fishbase-package"><i class="fa fa-check"></i><b>4.2.7</b> The Fishbase Package</a></li>
<li class="chapter" data-level="4.2.8" data-path="ch-04.html"><a href="ch-04.html#summary"><i class="fa fa-check"></i><b>4.2.8</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="ch-04.html"><a href="ch-04.html#case-study"><i class="fa fa-check"></i><b>4.3</b> Case Study</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="ch-04.html"><a href="ch-04.html#creating-list-of-species"><i class="fa fa-check"></i><b>4.3.1</b> Creating List of Species</a></li>
<li class="chapter" data-level="4.3.2" data-path="ch-04.html"><a href="ch-04.html#extracting-iucn-status-for-all-species"><i class="fa fa-check"></i><b>4.3.2</b> Extracting IUCN Status for all species</a></li>
<li class="chapter" data-level="4.3.3" data-path="ch-04.html"><a href="ch-04.html#proportion-of-species-in-netherlands"><i class="fa fa-check"></i><b>4.3.3</b> Proportion of species in Netherlands</a></li>
<li class="chapter" data-level="4.3.4" data-path="ch-04.html"><a href="ch-04.html#cleaning-data-with-iucn-status"><i class="fa fa-check"></i><b>4.3.4</b> Cleaning data with IUCN Status</a></li>
<li class="chapter" data-level="4.3.5" data-path="ch-04.html"><a href="ch-04.html#maps-of-species-richness-and-red-list-species-proportions"><i class="fa fa-check"></i><b>4.3.5</b> Maps of species richness and Red List species proportions</a></li>
<li class="chapter" data-level="4.3.6" data-path="ch-04.html"><a href="ch-04.html#relation-of-basin-size-and-species-richness"><i class="fa fa-check"></i><b>4.3.6</b> Relation of basin size and species richness</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="ch-04.html"><a href="ch-04.html#exercise-3"><i class="fa fa-check"></i><b>4.4</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="ch-05.html"><a href="ch-05.html"><i class="fa fa-check"></i><b>5</b> Catch-up</a>
<ul>
<li class="chapter" data-level="5.1" data-path="ch-05.html"><a href="ch-05.html#loops-in-r"><i class="fa fa-check"></i><b>5.1</b> Loops in R</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="ch-05.html"><a href="ch-05.html#some-simple-examples"><i class="fa fa-check"></i><b>5.1.1</b> Some simple examples</a></li>
<li class="chapter" data-level="5.1.2" data-path="ch-05.html"><a href="ch-05.html#nested-loops"><i class="fa fa-check"></i><b>5.1.2</b> Nested loops</a></li>
<li class="chapter" data-level="5.1.3" data-path="ch-05.html"><a href="ch-05.html#exercise-4"><i class="fa fa-check"></i><b>5.1.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="ch-05.html"><a href="ch-05.html#functional-programming-using-purr"><i class="fa fa-check"></i><b>5.2</b> Functional programming using purr</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="ch-05.html"><a href="ch-05.html#shortcuts-in-a-purrr-function"><i class="fa fa-check"></i><b>5.2.1</b> Shortcuts in a <code>purrr</code> function</a></li>
<li class="chapter" data-level="5.2.2" data-path="ch-05.html"><a href="ch-05.html#workflow-nested-data-map-and-mutate"><i class="fa fa-check"></i><b>5.2.2</b> Workflow: nested data, map and mutate</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="ch-05.html"><a href="ch-05.html#string-manipulations"><i class="fa fa-check"></i><b>5.3</b> String Manipulations</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="ch-05.html"><a href="ch-05.html#introduction-to-strings"><i class="fa fa-check"></i><b>5.3.1</b> Introduction to strings</a></li>
<li class="chapter" data-level="5.3.2" data-path="ch-05.html"><a href="ch-05.html#matching-and-extracting-patterns"><i class="fa fa-check"></i><b>5.3.2</b> Matching and extracting patterns</a></li>
<li class="chapter" data-level="5.3.3" data-path="ch-05.html"><a href="ch-05.html#advanced-example"><i class="fa fa-check"></i><b>5.3.3</b> Advanced example</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="ch-05.html"><a href="ch-05.html#web-scraping-in-a-nut-shell"><i class="fa fa-check"></i><b>5.4</b> Web-scraping in a nut-shell</a></li>
<li class="chapter" data-level="5.5" data-path="ch-05.html"><a href="ch-05.html#tidyverses-filter-and-select"><i class="fa fa-check"></i><b>5.5</b> Tidyverse’s filter and select</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="ch-05.html"><a href="ch-05.html#introduction-4"><i class="fa fa-check"></i><b>5.5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.5.2" data-path="ch-05.html"><a href="ch-05.html#select"><i class="fa fa-check"></i><b>5.5.2</b> Select()</a></li>
<li class="chapter" data-level="5.5.3" data-path="ch-05.html"><a href="ch-05.html#filter"><i class="fa fa-check"></i><b>5.5.3</b> Filter()</a></li>
<li class="chapter" data-level="5.5.4" data-path="ch-05.html"><a href="ch-05.html#exercises"><i class="fa fa-check"></i><b>5.5.4</b> Exercises</a></li>
<li class="chapter" data-level="5.5.5" data-path="ch-05.html"><a href="ch-05.html#solutions"><i class="fa fa-check"></i><b>5.5.5</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="ch-05.html"><a href="ch-05.html#preparing-data-for-ggplot"><i class="fa fa-check"></i><b>5.6</b> Preparing data for ggplot()</a></li>
<li class="chapter" data-level="5.7" data-path="ch-05.html"><a href="ch-05.html#base-r-functions"><i class="fa fa-check"></i><b>5.7</b> Base R functions</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="ch-06.html"><a href="ch-06.html"><i class="fa fa-check"></i><b>6</b> Supervised Machine Learning I</a>
<ul>
<li class="chapter" data-level="6.1" data-path="ch-06.html"><a href="ch-06.html#introduction-5"><i class="fa fa-check"></i><b>6.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="ch-06.html"><a href="ch-06.html#learning-objectives-1"><i class="fa fa-check"></i><b>6.1.1</b> Learning objectives</a></li>
<li class="chapter" data-level="6.1.2" data-path="ch-06.html"><a href="ch-06.html#important-points-from-the-lecture-1"><i class="fa fa-check"></i><b>6.1.2</b> Important points from the lecture</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="ch-06.html"><a href="ch-06.html#tutorial-4"><i class="fa fa-check"></i><b>6.2</b> Tutorial</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="ch-06.html"><a href="ch-06.html#linear-regression"><i class="fa fa-check"></i><b>6.2.1</b> Linear regression</a></li>
<li class="chapter" data-level="6.2.2" data-path="ch-06.html"><a href="ch-06.html#k-nearest-neighbours"><i class="fa fa-check"></i><b>6.2.2</b> K-nearest neighbours</a></li>
<li class="chapter" data-level="6.2.3" data-path="ch-06.html"><a href="ch-06.html#essential-methods-for-the-modelling-process"><i class="fa fa-check"></i><b>6.2.3</b> Essential methods for the modelling process</a></li>
<li class="chapter" data-level="6.2.4" data-path="ch-06.html"><a href="ch-06.html#bonus"><i class="fa fa-check"></i><b>6.2.4</b> Bonus</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="ch-06.html"><a href="ch-06.html#exercise-5"><i class="fa fa-check"></i><b>6.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="ch-07.html"><a href="ch-07.html"><i class="fa fa-check"></i><b>7</b> Supervised Machine Learning II</a>
<ul>
<li class="chapter" data-level="7.1" data-path="ch-07.html"><a href="ch-07.html#introduction-6"><i class="fa fa-check"></i><b>7.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="ch-07.html"><a href="ch-07.html#leaerning-objectives"><i class="fa fa-check"></i><b>7.1.1</b> Leaerning objectives</a></li>
<li class="chapter" data-level="7.1.2" data-path="ch-07.html"><a href="ch-07.html#key-points-from-the-lecture"><i class="fa fa-check"></i><b>7.1.2</b> Key points from the lecture</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="ch-07.html"><a href="ch-07.html#tutorial-5"><i class="fa fa-check"></i><b>7.2</b> Tutorial</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="ch-07.html"><a href="ch-07.html#model-training-and-the-loss-function"><i class="fa fa-check"></i><b>7.2.1</b> Model training and the loss function</a></li>
<li class="chapter" data-level="7.2.2" data-path="ch-07.html"><a href="ch-07.html#putting-it-all-together"><i class="fa fa-check"></i><b>7.2.2</b> Putting it all together</a></li>
<li class="chapter" data-level="7.2.3" data-path="ch-07.html"><a href="ch-07.html#model-evaluation"><i class="fa fa-check"></i><b>7.2.3</b> Model evaluation</a></li>
<li class="chapter" data-level="7.2.4" data-path="ch-07.html"><a href="ch-07.html#model-interpretation"><i class="fa fa-check"></i><b>7.2.4</b> Model interpretation</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="ch-07.html"><a href="ch-07.html#exercise-6"><i class="fa fa-check"></i><b>7.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="ch-08.html"><a href="ch-08.html"><i class="fa fa-check"></i><b>8</b> Application 1: Variable selection</a>
<ul>
<li class="chapter" data-level="8.1" data-path="ch-08.html"><a href="ch-08.html#introduction-7"><i class="fa fa-check"></i><b>8.1</b> Introduction</a></li>
<li class="chapter" data-level="8.2" data-path="ch-08.html"><a href="ch-08.html#application"><i class="fa fa-check"></i><b>8.2</b> Application</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="ch-08.html"><a href="ch-08.html#warm-up-1-nested-for-loop"><i class="fa fa-check"></i><b>8.2.1</b> Warm-up 1: Nested for-loop</a></li>
<li class="chapter" data-level="8.2.2" data-path="ch-08.html"><a href="ch-08.html#warm-up-2-find-the-best-single-predictor"><i class="fa fa-check"></i><b>8.2.2</b> Warm-up 2: Find the best single predictor</a></li>
<li class="chapter" data-level="8.2.3" data-path="ch-08.html"><a href="ch-08.html#full-stepwise-regression"><i class="fa fa-check"></i><b>8.2.3</b> Full stepwise regression</a></li>
<li class="chapter" data-level="8.2.4" data-path="ch-08.html"><a href="ch-08.html#bonus-stepwise-regression-out-of-the-box"><i class="fa fa-check"></i><b>8.2.4</b> Bonus: Stepwise regression out-of-the-box</a></li>
<li class="chapter" data-level="8.2.5" data-path="ch-08.html"><a href="ch-08.html#bonus-best-subset-selection"><i class="fa fa-check"></i><b>8.2.5</b> Bonus: Best Subset Selection</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="ch-09.html"><a href="ch-09.html"><i class="fa fa-check"></i><b>9</b> Supervised Neural Networks I</a>
<ul>
<li class="chapter" data-level="9.1" data-path="ch-09.html"><a href="ch-09.html#introduction-8"><i class="fa fa-check"></i><b>9.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="ch-09.html"><a href="ch-09.html#learning-objectives-2"><i class="fa fa-check"></i><b>9.1.1</b> Learning objectives</a></li>
<li class="chapter" data-level="9.1.2" data-path="ch-09.html"><a href="ch-09.html#important-points-from-the-lecture-2"><i class="fa fa-check"></i><b>9.1.2</b> Important points from the lecture</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="ch-09.html"><a href="ch-09.html#tutorial-6"><i class="fa fa-check"></i><b>9.2</b> Tutorial</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="ch-09.html"><a href="ch-09.html#set-up"><i class="fa fa-check"></i><b>9.2.1</b> Set-up</a></li>
<li class="chapter" data-level="9.2.2" data-path="ch-09.html"><a href="ch-09.html#keras-for-linear-models"><i class="fa fa-check"></i><b>9.2.2</b> Keras for linear models</a></li>
<li class="chapter" data-level="9.2.3" data-path="ch-09.html"><a href="ch-09.html#tuning-learning-rate"><i class="fa fa-check"></i><b>9.2.3</b> Tuning learning rate</a></li>
<li class="chapter" data-level="9.2.4" data-path="ch-09.html"><a href="ch-09.html#logistic-regression"><i class="fa fa-check"></i><b>9.2.4</b> Logistic Regression</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="ch-09.html"><a href="ch-09.html#exercise-7"><i class="fa fa-check"></i><b>9.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="ch-10.html"><a href="ch-10.html"><i class="fa fa-check"></i><b>10</b> Supervised Neural Networks II</a>
<ul>
<li class="chapter" data-level="10.1" data-path="ch-10.html"><a href="ch-10.html#introduction-9"><i class="fa fa-check"></i><b>10.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="ch-10.html"><a href="ch-10.html#learning-objectives-3"><i class="fa fa-check"></i><b>10.1.1</b> Learning objectives</a></li>
<li class="chapter" data-level="10.1.2" data-path="ch-10.html"><a href="ch-10.html#important-points-from-the-lecture-3"><i class="fa fa-check"></i><b>10.1.2</b> Important points from the lecture</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="ch-10.html"><a href="ch-10.html#tutorial-7"><i class="fa fa-check"></i><b>10.2</b> Tutorial</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="ch-10.html"><a href="ch-10.html#import-libraries-1"><i class="fa fa-check"></i><b>10.2.1</b> Import libraries</a></li>
<li class="chapter" data-level="10.2.2" data-path="ch-10.html"><a href="ch-10.html#construct-a-toy-dataset"><i class="fa fa-check"></i><b>10.2.2</b> Construct a toy dataset</a></li>
<li class="chapter" data-level="10.2.3" data-path="ch-10.html"><a href="ch-10.html#build-and-train-nn"><i class="fa fa-check"></i><b>10.2.3</b> Build and train NN</a></li>
<li class="chapter" data-level="10.2.4" data-path="ch-10.html"><a href="ch-10.html#model-performance"><i class="fa fa-check"></i><b>10.2.4</b> Model performance</a></li>
<li class="chapter" data-level="10.2.5" data-path="ch-10.html"><a href="ch-10.html#influence-of-nn-architecture"><i class="fa fa-check"></i><b>10.2.5</b> Influence of NN architecture</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="ch-10.html"><a href="ch-10.html#exercise-8"><i class="fa fa-check"></i><b>10.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="ch-11.html"><a href="ch-11.html"><i class="fa fa-check"></i><b>11</b> Application 2: Neural Networks and Hyperparameter Tuning</a>
<ul>
<li class="chapter" data-level="11.1" data-path="ch-11.html"><a href="ch-11.html#introduction-10"><i class="fa fa-check"></i><b>11.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="ch-11.html"><a href="ch-11.html#learning-goals"><i class="fa fa-check"></i><b>11.1.1</b> Learning Goals</a></li>
<li class="chapter" data-level="11.1.2" data-path="ch-11.html"><a href="ch-11.html#key-points-from-previous-lectures"><i class="fa fa-check"></i><b>11.1.2</b> Key Points from Previous Lectures</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="ch-11.html"><a href="ch-11.html#application-1"><i class="fa fa-check"></i><b>11.2</b> Application</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="ch-11.html"><a href="ch-11.html#problem-statement"><i class="fa fa-check"></i><b>11.2.1</b> Problem Statement</a></li>
<li class="chapter" data-level="11.2.2" data-path="ch-11.html"><a href="ch-11.html#data-preparation"><i class="fa fa-check"></i><b>11.2.2</b> Data preparation</a></li>
<li class="chapter" data-level="11.2.3" data-path="ch-11.html"><a href="ch-11.html#center-and-scale"><i class="fa fa-check"></i><b>11.2.3</b> Center and scale</a></li>
<li class="chapter" data-level="11.2.4" data-path="ch-11.html"><a href="ch-11.html#building-a-simple-model-with-keras-subtask-1"><i class="fa fa-check"></i><b>11.2.4</b> Building a simple model with keras ( SubTask 1)</a></li>
<li class="chapter" data-level="11.2.5" data-path="ch-11.html"><a href="ch-11.html#cross-validation"><i class="fa fa-check"></i><b>11.2.5</b> Cross validation</a></li>
<li class="chapter" data-level="11.2.6" data-path="ch-11.html"><a href="ch-11.html#parameter-tuning"><i class="fa fa-check"></i><b>11.2.6</b> Parameter Tuning</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="ch-12.html"><a href="ch-12.html"><i class="fa fa-check"></i><b>12</b> Supervised Deep Learning I</a>
<ul>
<li class="chapter" data-level="12.1" data-path="ch-12.html"><a href="ch-12.html#introduction-11"><i class="fa fa-check"></i><b>12.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="ch-12.html"><a href="ch-12.html#learning-objectives-4"><i class="fa fa-check"></i><b>12.1.1</b> Learning objectives</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="ch-12.html"><a href="ch-12.html#tutorial-8"><i class="fa fa-check"></i><b>12.2</b> Tutorial</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="ch-12.html"><a href="ch-12.html#building-blocks-of-cnns"><i class="fa fa-check"></i><b>12.2.1</b> Building Blocks of CNNs</a></li>
<li class="chapter" data-level="12.2.2" data-path="ch-12.html"><a href="ch-12.html#build-the-model"><i class="fa fa-check"></i><b>12.2.2</b> Build the model</a></li>
<li class="chapter" data-level="12.2.3" data-path="ch-12.html"><a href="ch-12.html#compile-and-train-the-model"><i class="fa fa-check"></i><b>12.2.3</b> Compile and train the model</a></li>
<li class="chapter" data-level="12.2.4" data-path="ch-12.html"><a href="ch-12.html#reduce-overfitting"><i class="fa fa-check"></i><b>12.2.4</b> Reduce Overfitting</a></li>
<li class="chapter" data-level="12.2.5" data-path="ch-12.html"><a href="ch-12.html#visualizing-a-cnn"><i class="fa fa-check"></i><b>12.2.5</b> Visualizing a CNN</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="ch-12.html"><a href="ch-12.html#exercise-9"><i class="fa fa-check"></i><b>12.3</b> Exercise</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="ch-12.html"><a href="ch-12.html#import-libraries-and-data"><i class="fa fa-check"></i><b>12.3.1</b> Import libraries and data</a></li>
<li class="chapter" data-level="12.3.2" data-path="ch-12.html"><a href="ch-12.html#tasks"><i class="fa fa-check"></i><b>12.3.2</b> Tasks</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="ch-13.html"><a href="ch-13.html"><i class="fa fa-check"></i><b>13</b> Supervised Deep Learning II</a>
<ul>
<li class="chapter" data-level="13.1" data-path="ch-13.html"><a href="ch-13.html#introduction-12"><i class="fa fa-check"></i><b>13.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="13.1.1" data-path="ch-13.html"><a href="ch-13.html#learning-objectives-5"><i class="fa fa-check"></i><b>13.1.1</b> Learning objectives</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="ch-13.html"><a href="ch-13.html#tutorial-9"><i class="fa fa-check"></i><b>13.2</b> Tutorial</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="ch-13.html"><a href="ch-13.html#dataset"><i class="fa fa-check"></i><b>13.2.1</b> Dataset</a></li>
<li class="chapter" data-level="13.2.2" data-path="ch-13.html"><a href="ch-13.html#naive-models-old-world"><i class="fa fa-check"></i><b>13.2.2</b> Naive models (Old World)</a></li>
<li class="chapter" data-level="13.2.3" data-path="ch-13.html"><a href="ch-13.html#neural-networks-new-world"><i class="fa fa-check"></i><b>13.2.3</b> Neural Networks (New World)</a></li>
<li class="chapter" data-level="13.2.4" data-path="ch-13.html"><a href="ch-13.html#model-comparison"><i class="fa fa-check"></i><b>13.2.4</b> Model Comparison</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="ch-13.html"><a href="ch-13.html#exercise-10"><i class="fa fa-check"></i><b>13.3</b> Exercise</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="ch-13.html"><a href="ch-13.html#import-libraries-2"><i class="fa fa-check"></i><b>13.3.1</b> Import libraries</a></li>
<li class="chapter" data-level="13.3.2" data-path="ch-13.html"><a href="ch-13.html#load-data"><i class="fa fa-check"></i><b>13.3.2</b> Load data</a></li>
<li class="chapter" data-level="13.3.3" data-path="ch-13.html"><a href="ch-13.html#preprocess-ndvi-images"><i class="fa fa-check"></i><b>13.3.3</b> Preprocess NDVI images</a></li>
<li class="chapter" data-level="13.3.4" data-path="ch-13.html"><a href="ch-13.html#part-1"><i class="fa fa-check"></i><b>13.3.4</b> Part 1</a></li>
<li class="chapter" data-level="13.3.5" data-path="ch-13.html"><a href="ch-13.html#part-2"><i class="fa fa-check"></i><b>13.3.5</b> Part 2</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Environmental Systems Data Science</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="ch-03" class="section level1" number="3">
<h1><span class="header-section-number">Chapter 3</span> Data variety</h1>
<div id="introduction-2" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Introduction</h2>
<div id="overview" class="section level3" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> Overview</h3>
<p>In this practical we will look at data variety in environmental sciences. We will start with the data from one eddy covariance measurement site in Switzerland encountered in Chapter 1, and complement it with remote sensing data. Specifically, we will download remote sensing data that measures vegetation greenness data, quantified by the normalised difference vegetation index, NDVI. We are interested in this data, because we expect ecosystem photosynthesis (gross primary production), measured at the eddy covariance tower, to covary with NDVI. Such covariation can provide powerful information for modelling. For example, we can train a machine learning model with the observed covariation between measured GPP and NDVI at eddy covariance sites and use that model to predict GPP from data obtained in space (NDVI is available for the whole globe, while GPP can only be measured locally.) You will be doing this kind of modelling in session 13. For now, we will focus on working with the spatial aspects of the data (since remote sensing data is inherently spatial) and introduce you to the tool set to work with spatial, in particular, geo-spatial data.</p>
<p>To get you familiar with spatial data types we will use a variety of data from freely accessible sites. We can obtain abiotic and biotic conditions at the tower locations. We will be spatially plotting the eddy towers you have already encountered in the first two chapters and extracting values for these locations from other data. By doing this we can show that the climate at these sites spans a gradient temperature and landcover. There are also biotic components that could explain variation in productivity among sites, including species composition and functional traits of plants.</p>
</div>
<div id="learning-objectives" class="section level3" number="3.1.2">
<h3><span class="header-section-number">3.1.2</span> Learning objectives</h3>
<p>After this learning unit, you will be able to …</p>
<ul>
<li>explain the possible sources of data in environmental sciences;</li>
<li>understand the range of operations applied to data from basic to complex;</li>
<li>read various types of data in R and apply basic operations;</li>
<li>understand the structure of spatial data, including rasters and shapefiles;</li>
<li>apply a range of operations to prepare data for analyses.</li>
</ul>
</div>
<div id="key-points-of-the-lecture" class="section level3" number="3.1.3">
<h3><span class="header-section-number">3.1.3</span> Key points of the lecture</h3>
<p>Data operations range in complexity, here they are listed from simplest to extremely complex:</p>
<ul>
<li>Statistical operations</li>
<li>Similarity metrics</li>
<li>Ordinations</li>
<li>Clustering</li>
<li>Classifications</li>
<li>Regressions</li>
<li>Neural network deep learning</li>
</ul>
<p>Examples of methods for the computing differences between datasets:</p>
<ul>
<li><em>Euclidean distance</em> is used for the simple quantification of variation</li>
<li><em>PCA</em> is a valuable tool to explore variation and reduce dimension</li>
<li><em>Hamming distance</em> helps measure the difference between strings</li>
<li><em>Jaccard index</em> describes the similarity between two or more binary data sets</li>
</ul>
<p>Machine learning is loosely defined as any learning done by a computer. It can be divided into…</p>
<ul>
<li><em>Unsupervised learning</em>: without explanatory data, which uses clustering to obtain a categorical output</li>
<li><em>Supervised learning</em>: with explanatory data, which uses either classification also resulting in a categorical output, or regression when a numerical output is required</li>
</ul>
<p>Data errors are divided into errors of accuracy and precision. If data is inaccurate, it strays from the true value. If data is imprecise, individual data points are variable under the same conditions. Systematic errors and inaccurate data is much harder to correct!</p>
<p>Errors can arise throughout the data science workflow.</p>
<ul>
<li><em>Inherent error</em> refers to the error present in the source documents and data.</li>
<li><em>Operational error</em> is created after data collection.</li>
</ul>
<p>Data can be cleaned from errors by looking for:</p>
<ul>
<li>Non-uniform data</li>
<li>Missing data</li>
<li>Duplicated data</li>
<li>Outliers</li>
<li>Measurement errors</li>
</ul>
<p>Quantifying uncertainty is extremely important, as undetectable errors will persist even after data cleaning. Examples of uncertainty quantification include…</p>
<ul>
<li>Confidence interval (CI)</li>
<li>Prediction interval</li>
</ul>
<p>Error propagation is used for datasets containing different uncertainties to determine the overall uncertainty.</p>
</div>
</div>
<div id="tutorial-2" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Tutorial</h2>
<div id="overview-1" class="section level3" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> Overview</h3>
<p>In this practical, you will be downloading and getting to know spatial data. There are three types of spatial data:</p>
<ul>
<li>points;</li>
<li>shapefiles;</li>
<li>rasters.</li>
</ul>
<p>The first part of this practical is mandatory and will start with a little introduction on downloading remote sensing data for our tower sites and seeing what kinds of information we can gather from this. Then we will get into the nitty-gritty aspects of the aforementioned spatial data types.</p>
<p>The second part is facultative, there we will discuss trait data and the principal component analysis (PCA), which is a method to reduce the dimensions of datasets.</p>
</div>
<div id="modis-remote-download" class="section level3" number="3.2.2">
<h3><span class="header-section-number">3.2.2</span> MODIS remote download</h3>
<p>We will be starting off this tutorial with an example on using remote sensing data. Let us focus on a fluxnet site already familiar to you: CH-Lae. The Lägern site is located on a mountain in the Swiss Plateau NW of Zürich and is surrounded by managed mixed deciduous mountain forests. The forest is highly diverse and dominated by beech. In the following figures, you can get an impression of the tower and its surroundings.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-122"></span>
<img src="figures/site1.jpg" alt="Images of Fluxnet sites and measurement station." width="32%" height="49%" /><img src="figures/site2.jpg" alt="Images of Fluxnet sites and measurement station." width="32%" height="49%" /><img src="figures/site3.jpg" alt="Images of Fluxnet sites and measurement station." width="32%" height="49%" />
<p class="caption">
Figure 3.1: Images of Fluxnet sites and measurement station.
</p>
</div>
<p>As always we start by loading <code>tidyverse</code>and our fluxnet site data.</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="ch-03.html#cb216-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb216-2"><a href="ch-03.html#cb216-2" aria-hidden="true" tabindex="-1"></a>df_sites <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;./data/fluxnet_site_info_reduced.csv&quot;</span>)</span></code></pre></div>
<p>Since we will only be working with the Lägern site for this section, we can extract its latitude and longitude.</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="ch-03.html#cb217-1" aria-hidden="true" tabindex="-1"></a>lon_lae <span class="ot">&lt;-</span> df_sites <span class="sc">%&gt;%</span> </span>
<span id="cb217-2"><a href="ch-03.html#cb217-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(site <span class="sc">==</span> <span class="st">&quot;CH-Lae&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb217-3"><a href="ch-03.html#cb217-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(lon)</span>
<span id="cb217-4"><a href="ch-03.html#cb217-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb217-5"><a href="ch-03.html#cb217-5" aria-hidden="true" tabindex="-1"></a>lat_lae <span class="ot">&lt;-</span> df_sites <span class="sc">%&gt;%</span> </span>
<span id="cb217-6"><a href="ch-03.html#cb217-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(site <span class="sc">==</span> <span class="st">&quot;CH-Lae&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb217-7"><a href="ch-03.html#cb217-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(lat)</span></code></pre></div>
<p>In this first analysis, we want to complement temporal data measured by the Fluxnet tower with remote sensing data. <em>Remote sensing</em> is the detection or monitoring of physical characteristics of the Earth’s surface by measuring reflected and emitted radiation at a distance. Data is collected in the form of images by special cameras typically from aircrafts or satellites. With this method, a huge amount of global data at large spatial scales gets easily accessible and therefore represents a useful method for the monitoring of ecosystems. However, this remotely acquired data requires validation from ground measurements, which can be done for example using the Eddy towers from the previous practicals.</p>
<p>One of these satellite remote sensing instruments is MODIS (Moderate Resolution Imaging Spectroradiometer). Terra MODIS and Aqua MODIS are viewing the entire Earth’s surface every 1 to 2 days, acquiring data in 36 spectral bands, or groups of wavelengths.</p>
<p>You used to have to visit an website to be able to download the data file by file, which was a long and tedious process. Luckily, now we have access to tools that make it much more straightforward to download remote sensing data from satellites. It provides Atmosphere, Ocean, Cryosphere and Land data series. The <code>MODISTools</code> package provides a programmatic interface to the <a href="https://modis.gsfc.nasa.gov/">MODIS Land Products Subsets web services</a> allows for easy downloads of “MODIS” time series (of single pixels or small regions of interest) directly to your R workspace or your computer.</p>
<ul>
<li>Using the <code>mt_...()</code> functions of the MODISTools we can check which products are available in MODIS, how the product we are searching for is called, and what temporal and spatial resolutions are available.</li>
<li><code>t_products()</code> lists all available products from MODIS with their temporal and spatial resolution. Products are parent categories of variables measured by the satellites, such as vegetation indices.</li>
<li><code>mt_bands()</code> requires the entry of a product chosen from “products” and list the available data variables that are represented by the different actually measured wavelengths.</li>
<li>Finally, <code>mt_dates()</code> lists all dates of which data from the chosen product and band are available.</li>
</ul>
<p>We will make use of this by loading NDVI data around our towers directly from MODIS.</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="ch-03.html#cb218-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MODISTools)</span>
<span id="cb218-2"><a href="ch-03.html#cb218-2" aria-hidden="true" tabindex="-1"></a>products <span class="ot">&lt;-</span> <span class="fu">mt_products</span>() <span class="sc">%&gt;%</span> <span class="fu">as_tibble</span>()</span>
<span id="cb218-3"><a href="ch-03.html#cb218-3" aria-hidden="true" tabindex="-1"></a>bands <span class="ot">&lt;-</span> <span class="fu">mt_bands</span>(<span class="at">product =</span> <span class="st">&quot;MOD13Q1&quot;</span>) <span class="sc">%&gt;%</span> <span class="fu">as_tibble</span>()</span>
<span id="cb218-4"><a href="ch-03.html#cb218-4" aria-hidden="true" tabindex="-1"></a>dates <span class="ot">&lt;-</span> <span class="fu">mt_dates</span>(<span class="at">product =</span> <span class="st">&quot;MOD13Q1&quot;</span>, <span class="at">lat =</span> lat_lae, <span class="at">lon =</span> lon_lae) <span class="sc">%&gt;%</span> <span class="fu">as_tibble</span>()</span></code></pre></div>
<p>Next, we will get the NDVI data for 6 years for a square of 2km x 2km around the tower site. We can use the function <code>mt_subset()</code> for this. The parameters of this function download a subset of data within the chosen product (here ‘MOD13Q1’). We specify the location as latitude and longitude, and the band we chose, which is at 250m spatial resolution and 16-day temporal resolution. The time period can be chosen as a subset of “dates” and must be provided in the form of start and end date. We are looking at the beginning of 2009 to the end of 2014. <code>km_lr</code> and <code>km_ab</code> define the kilometers to the left and right of the location and kilometers above and below, respectively. Since these values are being rounded to the nearest integer, we choose this minimum of 1, corresponding to 2km2. Sitename is used in writing data to file, internal gives the command whether the data should be returned as an internal structure. The progress setting defines whether the download progress should be shown or not.</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="ch-03.html#cb219-1" aria-hidden="true" tabindex="-1"></a>df_ndvi <span class="ot">&lt;-</span> <span class="fu">mt_subset</span>(<span class="at">product =</span> <span class="st">&quot;MOD13Q1&quot;</span>,            <span class="co"># the chosen product</span></span>
<span id="cb219-2"><a href="ch-03.html#cb219-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">lat =</span> lat_lae,                     <span class="co"># desired lat/lon</span></span>
<span id="cb219-3"><a href="ch-03.html#cb219-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">lon =</span> lon_lae,</span>
<span id="cb219-4"><a href="ch-03.html#cb219-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">band =</span> <span class="st">&quot;250m_16_days_NDVI&quot;</span>,        <span class="co"># chosen band defining spatial and temporal scale</span></span>
<span id="cb219-5"><a href="ch-03.html#cb219-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">start =</span> <span class="st">&quot;2009-01-01&quot;</span>,              <span class="co"># start date: 1st Jan 2009</span></span>
<span id="cb219-6"><a href="ch-03.html#cb219-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">end =</span> <span class="st">&quot;2014-12-19&quot;</span>,                <span class="co"># end date: 19th Dec 2014</span></span>
<span id="cb219-7"><a href="ch-03.html#cb219-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">km_lr =</span> <span class="dv">1</span>,                         <span class="co"># kilometers left &amp; right of the chosen location (lat/lon above)</span></span>
<span id="cb219-8"><a href="ch-03.html#cb219-8" aria-hidden="true" tabindex="-1"></a>                  <span class="at">km_ab =</span> <span class="dv">1</span>,                         <span class="co"># kilometers above and below the location</span></span>
<span id="cb219-9"><a href="ch-03.html#cb219-9" aria-hidden="true" tabindex="-1"></a>                  <span class="at">site_name =</span> <span class="st">&quot;CH-Lae&quot;</span>,              <span class="co"># the site name we want to give the data</span></span>
<span id="cb219-10"><a href="ch-03.html#cb219-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">internal =</span> <span class="cn">TRUE</span>,</span>
<span id="cb219-11"><a href="ch-03.html#cb219-11" aria-hidden="true" tabindex="-1"></a>                  <span class="at">progress =</span> <span class="cn">FALSE</span></span>
<span id="cb219-12"><a href="ch-03.html#cb219-12" aria-hidden="true" tabindex="-1"></a>                  ) <span class="sc">%&gt;%</span> </span>
<span id="cb219-13"><a href="ch-03.html#cb219-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>()</span></code></pre></div>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="ch-03.html#cb220-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_ndvi) <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">7</span>) <span class="co"># Only displaying 6 of 21 features</span></span></code></pre></div>
<pre><code>## # A tibble: 6 x 6
##   yllcorner  cellsize      nrows ncols band              units                
##   &lt;chr&gt;      &lt;chr&gt;         &lt;int&gt; &lt;int&gt; &lt;chr&gt;             &lt;chr&gt;                
## 1 5278290.12 231.656358264     9     9 250m_16_days_NDVI NDVI ratio - No units
## 2 5278290.12 231.656358264     9     9 250m_16_days_NDVI NDVI ratio - No units
## 3 5278290.12 231.656358264     9     9 250m_16_days_NDVI NDVI ratio - No units
## 4 5278290.12 231.656358264     9     9 250m_16_days_NDVI NDVI ratio - No units
## 5 5278290.12 231.656358264     9     9 250m_16_days_NDVI NDVI ratio - No units
## 6 5278290.12 231.656358264     9     9 250m_16_days_NDVI NDVI ratio - No units</code></pre>
<p>This dataframe takes some effort to make sense of. Let’s make our lives a little easier and put this data into a useful spatial context. For that it needs to be converted to a raster. This is a key spatial data type and will be described in more detail in the section Raster below.
The <code>MODISTools</code> packages provides a handy function, called <code>mt_to_raster()</code>, for converting such data frames produced by <code>mt_subset()</code> into a raster.</p>
<p>By plotting the data we can verify that we loaded a square of NDVI data around our tower for every 16-day time step.
We plot four images in the winter season.</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="ch-03.html#cb222-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(raster_ndvi[[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]], <span class="at">zlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.1</span>, <span class="fl">0.95</span>))</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-130-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>Just out of interest let’s calculate the minimum and maximum of these plots.</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="ch-03.html#cb223-1" aria-hidden="true" tabindex="-1"></a>min_winter <span class="ot">&lt;-</span> <span class="fu">min</span>(<span class="fu">minValue</span>(raster_ndvi[[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]]))</span>
<span id="cb223-2"><a href="ch-03.html#cb223-2" aria-hidden="true" tabindex="-1"></a>max_winter <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">maxValue</span>(raster_ndvi[[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]]))</span>
<span id="cb223-3"><a href="ch-03.html#cb223-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot; Minimum Winter: &quot;</span>, min_winter, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb223-4"><a href="ch-03.html#cb223-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Maximum Winter: &quot;</span>, max_winter)</span></code></pre></div>
<pre><code>##  Minimum Winter:  -0.06791774 
##  Maximum Winter:  0.6837171</code></pre>
<p>We do the same for the summer season.</p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="ch-03.html#cb225-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(raster_ndvi[[<span class="dv">13</span><span class="sc">:</span><span class="dv">16</span>]], <span class="at">zlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.1</span>, <span class="fl">0.95</span>))</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-132-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>We compute minimum and maximum.</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="ch-03.html#cb226-1" aria-hidden="true" tabindex="-1"></a>min_summer <span class="ot">&lt;-</span> <span class="fu">min</span>(<span class="fu">minValue</span>(raster_ndvi[[<span class="dv">13</span><span class="sc">:</span><span class="dv">16</span>]]))</span>
<span id="cb226-2"><a href="ch-03.html#cb226-2" aria-hidden="true" tabindex="-1"></a>max_summer <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">maxValue</span>(raster_ndvi[[<span class="dv">13</span><span class="sc">:</span><span class="dv">16</span>]]))</span>
<span id="cb226-3"><a href="ch-03.html#cb226-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot; Minimum Summer: &quot;</span>, min_summer, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb226-4"><a href="ch-03.html#cb226-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Maximum Summer: &quot;</span>, max_summer)</span></code></pre></div>
<pre><code>##  Minimum Summer:  0.569611 
##  Maximum Summer:  0.9034679</code></pre>
<p>From the two plots, we can clearly distinguish summer and winter time steps due to the differences in NDVI. Summer NDVI is much higher since our tower site is located in the Northern hemisphere where leaves get lost during winter for many vegetation types. Heterogeneity in winter is higher which might be due to conifers that don’t lose their leaves.</p>
<p>However, we are also interested in the mean across all (spatial) pixels for each date. To clarify this means we collapse the pixel dimension into a single mean value for each date. This is can done using <code>group_by()</code> in combination with <code>summarise()</code>, which were explained in Chapter <a href="ch-02.html#ch-02">2</a>.</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="ch-03.html#cb228-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lubridate) <span class="co"># lubridate is loaded to work with dates</span></span>
<span id="cb228-2"><a href="ch-03.html#cb228-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb228-3"><a href="ch-03.html#cb228-3" aria-hidden="true" tabindex="-1"></a><span class="do">## first determine the scaling factor that is to be applied to the NDVI values. This has practical reasons: reduces disk space</span></span>
<span id="cb228-4"><a href="ch-03.html#cb228-4" aria-hidden="true" tabindex="-1"></a>scale_factor <span class="ot">&lt;-</span> bands <span class="sc">%&gt;%</span> </span>
<span id="cb228-5"><a href="ch-03.html#cb228-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(band <span class="sc">==</span> <span class="st">&quot;250m_16_days_NDVI&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb228-6"><a href="ch-03.html#cb228-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(scale_factor) <span class="sc">%&gt;%</span> </span>
<span id="cb228-7"><a href="ch-03.html#cb228-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.numeric</span>()</span>
<span id="cb228-8"><a href="ch-03.html#cb228-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb228-9"><a href="ch-03.html#cb228-9" aria-hidden="true" tabindex="-1"></a>df_ndvi_spatialmean <span class="ot">&lt;-</span> df_ndvi <span class="sc">%&gt;%</span> </span>
<span id="cb228-10"><a href="ch-03.html#cb228-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">calendar_date =</span> <span class="fu">ymd</span>(calendar_date)) <span class="sc">%&gt;%</span>  <span class="co"># make the dates into comprehensible values not X.yyyy.mm.dd</span></span>
<span id="cb228-11"><a href="ch-03.html#cb228-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(calendar_date) <span class="sc">%&gt;%</span>                     <span class="co"># group the data by day </span></span>
<span id="cb228-12"><a href="ch-03.html#cb228-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean =</span> <span class="fu">mean</span>(value), <span class="at">min =</span> <span class="fu">min</span>(value), <span class="at">max =</span> <span class="fu">max</span>(value)) <span class="sc">%&gt;%</span>   <span class="co"># calculate mean, min and max across pixels</span></span>
<span id="cb228-13"><a href="ch-03.html#cb228-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mean =</span> mean <span class="sc">*</span> scale_factor, <span class="at">min =</span> min <span class="sc">*</span> scale_factor, <span class="at">max =</span> max <span class="sc">*</span> scale_factor)  <span class="co"># apply scale_factor (see )</span></span></code></pre></div>
<p>We can now plot our NDVI time series across our years (2009 to 2014) using the means we calculated above.</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="ch-03.html#cb229-1" aria-hidden="true" tabindex="-1"></a>df_ndvi_spatialmean <span class="sc">%&gt;%</span> </span>
<span id="cb229-2"><a href="ch-03.html#cb229-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> calendar_date)) <span class="sc">+</span></span>
<span id="cb229-3"><a href="ch-03.html#cb229-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> min, <span class="at">ymax =</span> max), <span class="at">fill =</span> <span class="st">&quot;grey70&quot;</span>) <span class="sc">+</span></span>
<span id="cb229-4"><a href="ch-03.html#cb229-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> mean))</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-135-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>After having aggregated the data to the mean NDVI across all 81 pixels for each date, we have reduced the dimensions of the data frame to only one (time). It now has a structure that can easily be combined with the time series data from the eddy covariance tower (each site is a point in space).</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="ch-03.html#cb230-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this can now be combined to the data frame with flux data that also has dates along rows</span></span>
<span id="cb230-2"><a href="ch-03.html#cb230-2" aria-hidden="true" tabindex="-1"></a>ddf_ch_lae <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;./data/ddf_ch_lae.csv&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="ch-03.html#cb231-1" aria-hidden="true" tabindex="-1"></a><span class="co"># now we can combine the two data frames along dates. </span></span>
<span id="cb231-2"><a href="ch-03.html#cb231-2" aria-hidden="true" tabindex="-1"></a>ddf_ch_lae_ndvi <span class="ot">&lt;-</span> ddf_ch_lae <span class="sc">%&gt;%</span></span>
<span id="cb231-3"><a href="ch-03.html#cb231-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">date =</span> TIMESTAMP) <span class="sc">%&gt;%</span> </span>
<span id="cb231-4"><a href="ch-03.html#cb231-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">year =</span> <span class="fu">year</span>(date)) <span class="sc">%&gt;%</span> </span>
<span id="cb231-5"><a href="ch-03.html#cb231-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(year <span class="sc">%in%</span> <span class="dv">2009</span><span class="sc">:</span><span class="dv">2014</span>) <span class="sc">%&gt;%</span>  <span class="co"># NDVI data was downloaded only for these years</span></span>
<span id="cb231-6"><a href="ch-03.html#cb231-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(df_ndvi_spatialmean <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">date =</span> calendar_date), <span class="at">by =</span> <span class="st">&quot;date&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb231-7"><a href="ch-03.html#cb231-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>year)</span></code></pre></div>
<p>NDVI data is provided every 16 days. In order to predict values at other time steps, we can fit a cubic smoothing spline to our daily values using <code>smooth.spline()</code>. In a nutshell, this is a way to fit a curve between two data points using cubic functions.</p>
<p>(In more detail, for the interested: Cubic smoothing splines embody a curve fitting technique that blends the ideas of cubic splines and curvature minimization to create an effective data modeling tool for noisy data. Traditional interpolating cubic splines represent the tabulated data as a piece-wise continuous curve which passes through each value in the data table. The curve spanning each data interval is represented by a cubic polynomial, with the requirement that the endpoints of adjacent cubic polynomials match in location and in their first and second derivatives. For more, have a look at this <a href="https://towardsdatascience.com/numerical-interpolation-natural-cubic-spline-52c1157b98ac">blogpost</a>)</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="ch-03.html#cb232-1" aria-hidden="true" tabindex="-1"></a>ddf_ch_lae_ndvi <span class="ot">&lt;-</span> ddf_ch_lae_ndvi <span class="sc">%&gt;%</span> </span>
<span id="cb232-2"><a href="ch-03.html#cb232-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">date_dec =</span> <span class="fu">decimal_date</span>(date))</span>
<span id="cb232-3"><a href="ch-03.html#cb232-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb232-4"><a href="ch-03.html#cb232-4" aria-hidden="true" tabindex="-1"></a>df_nona <span class="ot">&lt;-</span> ddf_ch_lae_ndvi <span class="sc">%&gt;%</span> </span>
<span id="cb232-5"><a href="ch-03.html#cb232-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>()</span>
<span id="cb232-6"><a href="ch-03.html#cb232-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb232-7"><a href="ch-03.html#cb232-7" aria-hidden="true" tabindex="-1"></a>out_spline <span class="ot">&lt;-</span> <span class="fu">smooth.spline</span>(df_nona<span class="sc">$</span>date_dec, df_nona<span class="sc">$</span>mean, <span class="at">spar=</span><span class="fl">0.1</span> )</span>
<span id="cb232-8"><a href="ch-03.html#cb232-8" aria-hidden="true" tabindex="-1"></a>vals_spline <span class="ot">&lt;-</span> <span class="fu">predict</span>(out_spline, ddf_ch_lae_ndvi<span class="sc">$</span>date_dec )<span class="sc">$</span>y</span>
<span id="cb232-9"><a href="ch-03.html#cb232-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb232-10"><a href="ch-03.html#cb232-10" aria-hidden="true" tabindex="-1"></a>ddf_ch_lae_ndvi <span class="ot">&lt;-</span> ddf_ch_lae_ndvi <span class="sc">%&gt;%</span> </span>
<span id="cb232-11"><a href="ch-03.html#cb232-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ndvi_splined =</span> vals_spline)</span></code></pre></div>
<p>Let’s take a look…</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="ch-03.html#cb233-1" aria-hidden="true" tabindex="-1"></a>ddf_ch_lae_ndvi <span class="sc">%&gt;%</span> </span>
<span id="cb233-2"><a href="ch-03.html#cb233-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb233-3"><a href="ch-03.html#cb233-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> date, <span class="at">y =</span> mean)) <span class="sc">+</span></span>
<span id="cb233-4"><a href="ch-03.html#cb233-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> date, <span class="at">y =</span> ndvi_splined), <span class="at">color =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-139-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>To see how well this data correlates with measurements from the fluxnet tower, we look at the relationship of GPP vs. NDVI*PPFD.</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="ch-03.html#cb234-1" aria-hidden="true" tabindex="-1"></a>ddf_ch_lae_ndvi <span class="ot">&lt;-</span> ddf_ch_lae_ndvi <span class="sc">%&gt;%</span> </span>
<span id="cb234-2"><a href="ch-03.html#cb234-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ppfd_abs =</span> ndvi_splined <span class="sc">*</span> PPFD_IN)</span></code></pre></div>
<p>Does NDVI affect the relationship between GPP and PPFD?</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="ch-03.html#cb235-1" aria-hidden="true" tabindex="-1"></a>linmod1 <span class="ot">&lt;-</span> <span class="fu">lm</span>(GPP_NT_VUT_REF <span class="sc">~</span> PPFD_IN, <span class="at">data =</span> ddf_ch_lae_ndvi)</span>
<span id="cb235-2"><a href="ch-03.html#cb235-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(linmod1)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = GPP_NT_VUT_REF ~ PPFD_IN, data = ddf_ch_lae_ndvi)
## 
## Residuals:
##      Min       1Q   Median       3Q      Max 
## -10.3749  -2.0413  -0.0137   1.9913  14.8576 
## 
## Coefficients:
##              Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept) 0.9348433  0.1097112   8.521   &lt;2e-16 ***
## PPFD_IN     0.0148270  0.0003038  48.797   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 3.06 on 2173 degrees of freedom
##   (16 observations deleted due to missingness)
## Multiple R-squared:  0.5229, Adjusted R-squared:  0.5226 
## F-statistic:  2381 on 1 and 2173 DF,  p-value: &lt; 2.2e-16</code></pre>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="ch-03.html#cb237-1" aria-hidden="true" tabindex="-1"></a>linmod2 <span class="ot">&lt;-</span> <span class="fu">lm</span>(GPP_NT_VUT_REF <span class="sc">~</span> ppfd_abs, <span class="at">data =</span> ddf_ch_lae_ndvi)</span>
<span id="cb237-2"><a href="ch-03.html#cb237-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(linmod2)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = GPP_NT_VUT_REF ~ ppfd_abs, data = ddf_ch_lae_ndvi)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -8.9202 -1.8307 -0.0829  1.7236 14.4273 
## 
## Coefficients:
##              Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept) 1.3395678  0.0903180   14.83   &lt;2e-16 ***
## ppfd_abs    0.0180868  0.0003148   57.45   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 2.792 on 2173 degrees of freedom
##   (16 observations deleted due to missingness)
## Multiple R-squared:  0.603,  Adjusted R-squared:  0.6028 
## F-statistic:  3301 on 1 and 2173 DF,  p-value: &lt; 2.2e-16</code></pre>
<p>Indeed, R2 increased from 0.5229 to 0.603 when including considering NDVI * PPFD, instead of just PPFD. In other words, it also matters whether leaves are actually out and green for modelling the relationship between GPP and PPFD. NDVI * PPFD approximates the amount of <em>absorbed</em> incoming light (not just the incoming light) and is therefore closer to being a physiologically relevant quantity.</p>
</div>
<div id="points-on-the-globe" class="section level3" number="3.2.3">
<h3><span class="header-section-number">3.2.3</span> Points on the globe</h3>
<p>Documenting the coordinates of data as it is collected allows us to assign a precise location to that data, which can be important in further analysis when, for example, comparing location. As you all undoubtedly know coordinates come with a latitude (defining the position N or S of the equator) and longitude (defining the position E or W of the meridian). With the coordinates we can already plot our data points, remembering that the latitude is equivalent to the y-axis and longitude to the x-axis.</p>
<p>In this section, we look at the position of the Fluxnet sites. The Fluxnet network measures land-atmosphere exchanges of greenhouse gases and energy for sites across the globe using the eddy covariance technique. High-frequency measurements of vertical wind velocity and a scalar mixing ratio (CO2, H2O, temperature, etc.) provides estimates of the net exchange of the scalar. Eddy covariance is currently the standard method to measure fluxes of trace gases between ecosystems and the atmosphere (sources:
<a href="https://fluxnet.org/about/">Fluxnet</a>, <a href="https://www.nature.com/articles/s41597-020-0534-3.pdf">Nature Article on Fluxnet Dataset</a>.</p>
<p>Right at the beginning of this tutorial we loaded the dataset containing information on sites in Europe. We treat site locations as a geographical position. So we will extract the name, longitude and latitude of our sites.</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="ch-03.html#cb239-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_sites)</span></code></pre></div>
<pre><code>## # A tibble: 6 x 3
##   site     lon   lat
##   &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt;
## 1 BE-Bra  4.52  51.3
## 2 BE-Lon  4.75  50.6
## 3 BE-Vie  6.00  50.3
## 4 CH-Cha  8.41  47.2
## 5 CH-Fru  8.54  47.1
## 6 CH-Lae  8.36  47.5</code></pre>
<p>Our dataset <em>‘df_sites’</em> contains two sites in Greenland and Siberia. We’ll focus on a smaller spatial domain, restricted to “mainland” Europe, therefore, we exclude sites RU-Cok and DK-ZaH.</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="ch-03.html#cb241-1" aria-hidden="true" tabindex="-1"></a>df_sites <span class="ot">&lt;-</span> df_sites <span class="sc">%&gt;%</span> </span>
<span id="cb241-2"><a href="ch-03.html#cb241-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span>(site <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;RU-Cok&quot;</span>, <span class="st">&quot;DK-ZaH&quot;</span>)))</span></code></pre></div>
<p>Already now we can plot our data points without any further adjustments.</p>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="ch-03.html#cb242-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb242-2"><a href="ch-03.html#cb242-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> df_sites, <span class="fu">aes</span>(<span class="at">x =</span> lon, <span class="at">y =</span> lat), <span class="at">color =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb242-3"><a href="ch-03.html#cb242-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Selected FLUXNET sites&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Longitude&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Latitude&quot;</span>)<span class="sc">+</span></span>
<span id="cb242-4"><a href="ch-03.html#cb242-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-145-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>We can make an educated guess where the points are but this plot doesn’t tell us much by itself without a map as a reference.</p>
<p>So to visualize the location of our points we plot them on a map. We use the package <code>SpData</code> to get the world map and crop it to Europe.</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="ch-03.html#cb243-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sf)</span>
<span id="cb243-2"><a href="ch-03.html#cb243-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(spData)</span>
<span id="cb243-3"><a href="ch-03.html#cb243-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb243-4"><a href="ch-03.html#cb243-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb243-5"><a href="ch-03.html#cb243-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> world) <span class="sc">+</span></span>
<span id="cb243-6"><a href="ch-03.html#cb243-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> df_sites, <span class="fu">aes</span>(<span class="at">x =</span> lon, <span class="at">y =</span> lat), <span class="at">color =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb243-7"><a href="ch-03.html#cb243-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Longitude&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Latitude&quot;</span>) <span class="sc">+</span></span>
<span id="cb243-8"><a href="ch-03.html#cb243-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_sf</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">30</span>,<span class="dv">40</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">35</span>,<span class="dv">80</span>), <span class="at">expand =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb243-9"><a href="ch-03.html#cb243-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-146-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>Of course, we can carry on using the data points as they are (for plotting or analysis) but for more detailed spatial analysis we will turn them into <em>SpatialPoints</em>.</p>
<p><em>SpatialPoints</em> consist of a matrix with <em>n</em> rows and 2 columns, one for each coordinate (latitude &amp; longitude). <em>n</em> is the number of points in the data. These points also have a so-called <em>‘projection string’</em> or <em>‘crs’</em> indicating the coordinate reference system in which coordinates of points are expressed.
There are different ways (formulas) to project the earth (an ellipsoid) onto a 2-dimensional map. You can only perform calculations, if the same method (projection, coordinate reference system (crs)) is used. Having points in different coordinate reference systems can cause data to look very skewed. It’s good to know which coordinate reference system your data is in, the most commonly used one is <strong>WGS84 (EPSG: 4326)</strong>. There are some useful resources to convert help you find the correct <a href="https://spatialreference.org/ref/epsg/4326/">spatial reference</a>.</p>
<p>To transform points into <em>SpatialPoints</em> we use the function <code>SpatialPoints</code> in the r package <code>sp</code>.</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="ch-03.html#cb244-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sp)</span>
<span id="cb244-2"><a href="ch-03.html#cb244-2" aria-hidden="true" tabindex="-1"></a>sp_sites <span class="ot">&lt;-</span> <span class="fu">SpatialPoints</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(df_sites, <span class="sc">-</span>site)) <span class="co"># remove character column</span></span></code></pre></div>
<p>Below you can see how with the R-base call <code>plot()</code> you can plot the <em>SpatialPoints</em> directly. But we’ll have to also provide a map to plot them onto. The loaded shapefile of europe will also be used in the next part called Shapefiles.</p>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="ch-03.html#cb245-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rgdal)</span>
<span id="cb245-2"><a href="ch-03.html#cb245-2" aria-hidden="true" tabindex="-1"></a>europe_shape <span class="ot">&lt;-</span> <span class="fu">readOGR</span>(<span class="at">dsn=</span><span class="st">&quot;./data/shapefiles&quot;</span>, <span class="at">layer=</span><span class="st">&quot;europe_map&quot;</span>)</span></code></pre></div>
<pre><code>## OGR data source with driver: ESRI Shapefile 
## Source: &quot;/Users/benjaminstocker/polybox/Shared/Data Science Lecture Planning - shared folder/4 Datasets/shapefiles&quot;, layer: &quot;europe_map&quot;
## with 53 features
## It has 94 fields</code></pre>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="ch-03.html#cb247-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(europe_shape, <span class="at">col=</span><span class="st">&quot;grey&quot;</span>)</span>
<span id="cb247-2"><a href="ch-03.html#cb247-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sp_sites, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>, <span class="at">add =</span> <span class="cn">TRUE</span>, <span class="at">pch =</span> <span class="dv">16</span>)</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-148-1.png" width="50%" style="display: block; margin: auto;" /></p>
</div>
<div id="shapefiles" class="section level3" number="3.2.4">
<h3><span class="header-section-number">3.2.4</span> Shapefiles</h3>
<p>Above we looked at points on the globe and transformed our data to <em>SpatialPoints</em>. This was already the first example of what forms a shapefile can come in. Shapefiles basically store geographic information, meaning location and any additional information, in shape objects. Shape objects in R are defined by <em>SpatialPoints</em>, <em>SpatialLines</em>, and <em>SpatialPolygons</em> classes of the <code>sp</code> package. The corresponding <em>SpatialPointsDataFrame</em>, <em>SpatialLinesDataFrame</em>, and <em>SpatialPolygonsDataFrame</em> classes allow storing shape objects together with a dataframe. The number of rows in the dataframe corresponds to the number of points (number of rows in coordinate matrix), lines (size of the list of Lines), or polygons (size of the list of Polygons). This allows us to associate a vector of variables (a row of the dataframe) to each shape object.
In the next step, we load a shapefile containing the countries of the world provided by <a href="www.naturalearthdata.com">Natural Earth</a>, as a reference with which we can contextualize our spatial data. Since we only are looking at towers in Europe, we have cropped the raster to Europe for you.</p>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="ch-03.html#cb248-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rgdal)</span>
<span id="cb248-2"><a href="ch-03.html#cb248-2" aria-hidden="true" tabindex="-1"></a>europe_shape <span class="ot">&lt;-</span> <span class="fu">readOGR</span>(<span class="at">dsn=</span><span class="st">&quot;./data/shapefiles&quot;</span>, <span class="at">layer=</span><span class="st">&quot;europe_map&quot;</span>)</span></code></pre></div>
<pre><code>## OGR data source with driver: ESRI Shapefile 
## Source: &quot;/Users/benjaminstocker/polybox/Shared/Data Science Lecture Planning - shared folder/4 Datasets/shapefiles&quot;, layer: &quot;europe_map&quot;
## with 53 features
## It has 94 fields</code></pre>
<p>Let’s see what class <em>europe_shape</em> is:</p>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="ch-03.html#cb250-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(europe_shape)</span></code></pre></div>
<pre><code>## [1] &quot;SpatialPolygonsDataFrame&quot;
## attr(,&quot;package&quot;)
## [1] &quot;sp&quot;</code></pre>
<p>Above we mentioned that <em>SpatialPolygonsDataFrame</em> contains both a list of <em>SpatialPolygons</em> and a dataframe with information on each polygon. Here’s how you can access the information in the dataframe. We’ll just display the header of some of the 94 features for now.</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="ch-03.html#cb252-1" aria-hidden="true" tabindex="-1"></a>europe_shape<span class="sc">@</span>data <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="fu">c</span>(NAME, SOVEREIGNT, SUBREGION, POP_EST))<span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code></pre></div>
<pre><code>##             NAME     SOVEREIGNT       SUBREGION  POP_EST
## 0        Vatican        Vatican Southern Europe     1000
## 1         Jersey United Kingdom Northern Europe    98840
## 2       Guernsey United Kingdom Northern Europe    66502
## 3    Isle of Man United Kingdom Northern Europe    88815
## 4 United Kingdom United Kingdom Northern Europe 64769452
## 5        Ukraine        Ukraine  Eastern Europe 44033874</code></pre>
<p>Let’s first plot our towers on this map of Europe. Remember, how when you transformed the tower locations we mentioned each point has a designated coordinate reference system? Here, we need to extract this projection from <em>europe_shape</em> using <code>proj4string()</code> and add it to our <em>SpatialPoints</em>. This way we can make sure they are in the same reference system for plotting them together.</p>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="ch-03.html#cb254-1" aria-hidden="true" tabindex="-1"></a>geo.proj <span class="ot">&lt;-</span> sp<span class="sc">::</span><span class="fu">proj4string</span>(europe_shape)</span>
<span id="cb254-2"><a href="ch-03.html#cb254-2" aria-hidden="true" tabindex="-1"></a>pts <span class="ot">&lt;-</span> sp<span class="sc">::</span><span class="fu">SpatialPoints</span>(sp_sites, <span class="at">proj4string =</span> sp<span class="sc">::</span><span class="fu">CRS</span>(geo.proj))</span></code></pre></div>
<p>Then we can take the <em>SpatialPoints</em>, now in the correct coordinate reference system, the data from <em>europe_shape</em> and bind them to our tower sites in the dataframe <em>df_sites</em>.</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="ch-03.html#cb255-1" aria-hidden="true" tabindex="-1"></a>df_sites_country_info <span class="ot">&lt;-</span> sp<span class="sc">::</span><span class="fu">over</span>(pts, europe_shape) <span class="sc">%&gt;%</span> </span>
<span id="cb255-2"><a href="ch-03.html#cb255-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb255-3"><a href="ch-03.html#cb255-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">bind_cols</span>(df_sites, .)</span>
<span id="cb255-4"><a href="ch-03.html#cb255-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb255-5"><a href="ch-03.html#cb255-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_sites_country_info) <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)</span></code></pre></div>
<pre><code>## # A tibble: 6 x 6
##   site     lon   lat featurecla      scalerank LABELRANK
##   &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;               &lt;int&gt;     &lt;int&gt;
## 1 BE-Bra  4.52  51.3 Admin-0 country         1         2
## 2 BE-Lon  4.75  50.6 Admin-0 country         1         2
## 3 BE-Vie  6.00  50.3 Admin-0 country         1         2
## 4 CH-Cha  8.41  47.2 Admin-0 country         1         4
## 5 CH-Fru  8.54  47.1 Admin-0 country         1         4
## 6 CH-Lae  8.36  47.5 Admin-0 country         1         4</code></pre>
<p>For the next part we will be using a map European with country boarders. For this, we’ll make an object called <em>shp_df</em>, this we do using the column <em>SOVEREIGNT</em>, which we saw when looking at <code>europe_shape@data</code>. This preserves only the data we need for further analysis and plotting.</p>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="ch-03.html#cb257-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(maptools)</span>
<span id="cb257-2"><a href="ch-03.html#cb257-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-3"><a href="ch-03.html#cb257-3" aria-hidden="true" tabindex="-1"></a>shp_df <span class="ot">&lt;-</span> broom<span class="sc">::</span><span class="fu">tidy</span>(europe_shape, <span class="at">region =</span> <span class="st">&quot;SOVEREIGNT&quot;</span>)</span>
<span id="cb257-4"><a href="ch-03.html#cb257-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(shp_df)</span></code></pre></div>
<pre><code>## # A tibble: 6 x 7
##    long   lat order hole  piece group     id     
##   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;lgl&gt; &lt;fct&gt; &lt;fct&gt;     &lt;chr&gt;  
## 1  19.3  41.9     1 FALSE 1     Albania.1 Albania
## 2  19.3  41.9     2 FALSE 1     Albania.1 Albania
## 3  19.4  42.0     3 FALSE 1     Albania.1 Albania
## 4  19.4  42.0     4 FALSE 1     Albania.1 Albania
## 5  19.4  42.1     5 FALSE 1     Albania.1 Albania
## 6  19.3  42.1     6 FALSE 1     Albania.1 Albania</code></pre>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="ch-03.html#cb259-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb259-2"><a href="ch-03.html#cb259-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_polygon</span>(<span class="at">data =</span> shp_df, <span class="fu">aes</span>(<span class="at">x =</span> long, <span class="at">y =</span> lat, <span class="at">group =</span> group, <span class="at">fill =</span> id), <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb259-3"><a href="ch-03.html#cb259-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb259-4"><a href="ch-03.html#cb259-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-154-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>As we saw in the earlier plot of the tower locations only some countries in Europe contain towers. Therefore, we want to crop our map to include only those countries.</p>
<p>To do this we first make a vector which is a list of the country each tower is located in. We get this from <em>df_site_country_info</em>.
Then we filter out only those countries with a tower in them and voilà, plot the countries and tower locations as a result.</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="ch-03.html#cb260-1" aria-hidden="true" tabindex="-1"></a>countries_with_site <span class="ot">&lt;-</span> df_sites_country_info <span class="sc">%&gt;%</span> <span class="fu">pull</span>(SOVEREIGNT) <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span>
<span id="cb260-2"><a href="ch-03.html#cb260-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-3"><a href="ch-03.html#cb260-3" aria-hidden="true" tabindex="-1"></a><span class="co"># resulting character vector (without NAs)</span></span>
<span id="cb260-4"><a href="ch-03.html#cb260-4" aria-hidden="true" tabindex="-1"></a>countries_with_site <span class="sc">%&gt;%</span> <span class="fu">na.omit</span>()</span></code></pre></div>
<pre><code>##  [1] &quot;Belgium&quot;     &quot;Belgium&quot;     &quot;Belgium&quot;     &quot;Switzerland&quot; &quot;Switzerland&quot;
##  [6] &quot;Switzerland&quot; &quot;Switzerland&quot; &quot;Switzerland&quot; &quot;Switzerland&quot; &quot;Germany&quot;    
## [11] &quot;Germany&quot;     &quot;Germany&quot;     &quot;Germany&quot;     &quot;Germany&quot;     &quot;Denmark&quot;    
## [16] &quot;Finland&quot;     &quot;Finland&quot;     &quot;France&quot;      &quot;France&quot;      &quot;France&quot;     
## [21] &quot;France&quot;      &quot;Italy&quot;       &quot;Italy&quot;       &quot;Italy&quot;       &quot;Italy&quot;      
## [26] &quot;Italy&quot;       &quot;Italy&quot;       &quot;Italy&quot;       &quot;Italy&quot;       &quot;Italy&quot;      
## [31] &quot;Netherlands&quot; &quot;Czechia&quot;    
## attr(,&quot;na.action&quot;)
## [1] 33
## attr(,&quot;class&quot;)
## [1] &quot;omit&quot;</code></pre>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="ch-03.html#cb262-1" aria-hidden="true" tabindex="-1"></a><span class="co"># country with towers filtered</span></span>
<span id="cb262-2"><a href="ch-03.html#cb262-2" aria-hidden="true" tabindex="-1"></a>shp_df_sub <span class="ot">&lt;-</span> shp_df <span class="sc">%&gt;%</span> </span>
<span id="cb262-3"><a href="ch-03.html#cb262-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(id <span class="sc">%in%</span> countries_with_site)</span>
<span id="cb262-4"><a href="ch-03.html#cb262-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb262-5"><a href="ch-03.html#cb262-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot result</span></span>
<span id="cb262-6"><a href="ch-03.html#cb262-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb262-7"><a href="ch-03.html#cb262-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_polygon</span>(<span class="at">data =</span> shp_df_sub, <span class="fu">aes</span>(<span class="at">x =</span> long, <span class="at">y =</span> lat, <span class="at">group =</span> group, <span class="at">fill =</span> id), <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb262-8"><a href="ch-03.html#cb262-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> df_sites, <span class="fu">aes</span>(<span class="at">x =</span> lon, <span class="at">y =</span> lat)) <span class="sc">+</span></span>
<span id="cb262-9"><a href="ch-03.html#cb262-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-155-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p><strong>Checkpoint</strong></p>
<p>You just learnt how to plot <em>europe_shape</em> coloured by country (“SOVREIGNT”), now plot it coloured by region in europe.
Start by finding the column that gives you this information in <code>europe_shape@data</code>.</p>
<p><strong>Solution</strong></p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="ch-03.html#cb263-1" aria-hidden="true" tabindex="-1"></a><span class="co"># take a look at the first 6 column names</span></span>
<span id="cb263-2"><a href="ch-03.html#cb263-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(europe_shape<span class="sc">@</span>data)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;featurecla&quot; &quot;scalerank&quot;  &quot;LABELRANK&quot;  &quot;SOVEREIGNT&quot; &quot;SOV_A3&quot;    
##  [6] &quot;ADM0_DIF&quot;   &quot;LEVEL&quot;      &quot;TYPE&quot;       &quot;ADMIN&quot;      &quot;ADM0_A3&quot;</code></pre>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="ch-03.html#cb265-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a df of the data needed from europe_shape</span></span>
<span id="cb265-2"><a href="ch-03.html#cb265-2" aria-hidden="true" tabindex="-1"></a>shp_df_region <span class="ot">&lt;-</span> broom<span class="sc">::</span><span class="fu">tidy</span>(europe_shape, <span class="at">region =</span> <span class="st">&quot;SUBREGION&quot;</span>)</span>
<span id="cb265-3"><a href="ch-03.html#cb265-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(shp_df_region)</span></code></pre></div>
<pre><code>## # A tibble: 6 x 7
##    long   lat order hole  piece group            id            
##   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;lgl&gt; &lt;fct&gt; &lt;fct&gt;            &lt;chr&gt;         
## 1  28    42.0     1 FALSE 1     Eastern Europe.1 Eastern Europe
## 2  27.9  42.0     2 FALSE 1     Eastern Europe.1 Eastern Europe
## 3  27.8  42.0     3 FALSE 1     Eastern Europe.1 Eastern Europe
## 4  27.8  42.0     4 FALSE 1     Eastern Europe.1 Eastern Europe
## 5  27.7  42.0     5 FALSE 1     Eastern Europe.1 Eastern Europe
## 6  27.7  42.0     6 FALSE 1     Eastern Europe.1 Eastern Europe</code></pre>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="ch-03.html#cb267-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot europe coloured by region</span></span>
<span id="cb267-2"><a href="ch-03.html#cb267-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb267-3"><a href="ch-03.html#cb267-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_polygon</span>(<span class="at">data =</span> shp_df_region, <span class="fu">aes</span>(<span class="at">x =</span> long, <span class="at">y =</span> lat, <span class="at">group =</span> group, <span class="at">fill =</span> id), <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb267-4"><a href="ch-03.html#cb267-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> df_sites, <span class="fu">aes</span>(<span class="at">x =</span> lon, <span class="at">y =</span> lat)) <span class="sc">+</span></span>
<span id="cb267-5"><a href="ch-03.html#cb267-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-156-1.png" width="50%" style="display: block; margin: auto;" /></p>
</div>
<div id="rasters" class="section level3" number="3.2.5">
<h3><span class="header-section-number">3.2.5</span> Rasters</h3>
<p>Next up are rasters. A raster is a matrix or grid of cells each of which contains information in the form of a value.</p>
<p>A raster object consists primarily of:</p>
<ul>
<li>A grid of cells;<br />
</li>
<li>A coordinate reference system (CRS) for the grid and its cells so that we know the location to which the grid refers;</li>
<li>A variable of interest for which each cell in the grid has a value, and;</li>
<li>Other information relating to the CRS, projection, resolution, etc.</li>
</ul>
<p>Let’s take a look at some rasters and get familiar with some functions to analyse them.</p>
<p>Since our goal ultimate goal is to predict productivity, we will consider different factors that might explain productivity, for example, landcover and temperature.
To start off we’ll look at some landcover data. With this data, we can investigate the surroundings of the Fluxnet towers. This data gives us the different classes of physical coverage of the Earth’s surface, such as forests, grasslands, croplands, lakes, etc. Landcover types affect fluxes measured by the tower through the vegetation that characterizes them, or by influencing the radiation budget through albedo, or water availability. Although we have some information from the overview file of the towers, we can still check whether we can confirm this information. We will load landcover data from GlobCover provided by the <a href="http://due.esrin.esa.int/page_globcover.php">European space agency (ESA)</a>. This project provides landcover maps observations of surface reflectance from the 300m MERIS sensor on board the ENVISAT satellite mission as their input.</p>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="ch-03.html#cb268-1" aria-hidden="true" tabindex="-1"></a><span class="co"># library(raster) -&gt; we have already done this but remember you need this to load rasters!</span></span>
<span id="cb268-2"><a href="ch-03.html#cb268-2" aria-hidden="true" tabindex="-1"></a>raster_landcover <span class="ot">&lt;-</span> <span class="fu">raster</span>(<span class="st">&quot;./data/Globcover_EU.tif&quot;</span>)</span></code></pre></div>
<p>This landcover data consists of different categories, these GlobCover categories are based on the Land Cover Classification System (LCCS) which was
developed by the Food and Agricultural Organization of the United Nations (FAO) to provide a consistent framework for the classification of mapping land cover. They include various types of forest, shrub- and grasslands, cropland and artificial surfaces.</p>
<p>Before we look at the data, we first reduce the spatial extent that the raster covers. We crop it to a rectangle around the site location of CH-Lae (+/- 2 degrees in longitudinal direction, +/- 1 degree in latitudinal direction).</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="ch-03.html#cb269-1" aria-hidden="true" tabindex="-1"></a>bounding_box <span class="ot">&lt;-</span> <span class="fu">extent</span>(lon_lae<span class="dv">-2</span>, lon_lae<span class="sc">+</span><span class="dv">2</span>, lat_lae<span class="dv">-1</span>, lat_lae<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb269-2"><a href="ch-03.html#cb269-2" aria-hidden="true" tabindex="-1"></a>raster_landcover_crop <span class="ot">&lt;-</span> <span class="fu">crop</span>(raster_landcover, bounding_box)</span></code></pre></div>
<p>We can plot our raster as the region around our tower. To indicate the position of our tower, we add a red dot.</p>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="ch-03.html#cb270-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(raster_landcover_crop, <span class="at">legend=</span><span class="cn">FALSE</span>, <span class="at">xlab=</span><span class="st">&quot;longitude&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;latitude&quot;</span>)</span>
<span id="cb270-2"><a href="ch-03.html#cb270-2" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(lon_lae, lat_lae, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p>If we want to conduct analysis using two different rasters, we have to make sure they have the same resolution and are aligned on the same grid. Let’s read a second raster and compare the grids. We load spatio-temporal temperature rasters available from <a href="https://chelsa-climate.org">CHELSA</a> to complement our flux tower measurements for further analysis.
The temperature raster data from CHELSA provides free high-resolution climate data (temperature and precipitation) for the past and the future. The past data we use here are downscaled model output temperature and precipitation estimates of the ERA-Interim climatic reanalysis (forecast models and data assimilation systems reanalysing archived observations). While the temperature algorithm is based on statistical downscaling of atmospheric temperatures, the precipitation algorithm incorporates orographic predictors including wind fields, valley exposition, and boundary layer height, with subsequent bias correction. The data we use consists of a monthly temperature and precipitation time series over the area of Europe from 2006 to 2012 in January. To compare this new raster with the previous one we need to crop it to the same region as the tower.</p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb271-1"><a href="ch-03.html#cb271-1" aria-hidden="true" tabindex="-1"></a>raster_chelsa <span class="ot">&lt;-</span> <span class="fu">raster</span>(<span class="st">&quot;./data/Chelsa_t_mean_2006-2012.tif&quot;</span>)</span>
<span id="cb271-2"><a href="ch-03.html#cb271-2" aria-hidden="true" tabindex="-1"></a>raster_chelsa_crop <span class="ot">&lt;-</span> <span class="fu">crop</span>(raster_chelsa, bounding_box)</span></code></pre></div>
<p>We now plot our raster.</p>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="ch-03.html#cb272-1" aria-hidden="true" tabindex="-1"></a>pal <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;purple&quot;</span>,<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;cyan&quot;</span>,<span class="st">&quot;green&quot;</span>,<span class="st">&quot;yellow&quot;</span>,<span class="st">&quot;red&quot;</span>))</span>
<span id="cb272-2"><a href="ch-03.html#cb272-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(raster_chelsa_crop, <span class="at">col =</span> <span class="fu">pal</span>(<span class="dv">20</span>), <span class="at">xlab=</span><span class="st">&quot;longitude&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;latitude&quot;</span>)</span>
<span id="cb272-3"><a href="ch-03.html#cb272-3" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(lon_lae, lat_lae, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-161-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>We see that the temperature units are not in a comprehensive format. To change the unit from Kelvin x10 to degree Celsius we have to divide the data by 10 and subtract 273.15.</p>
<div class="sourceCode" id="cb273"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb273-1"><a href="ch-03.html#cb273-1" aria-hidden="true" tabindex="-1"></a>raster_chelsa_crop <span class="ot">&lt;-</span> raster_chelsa_crop<span class="sc">/</span><span class="dv">10</span><span class="fl">-273.15</span></span></code></pre></div>
<p>We plot our raster again.</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="ch-03.html#cb274-1" aria-hidden="true" tabindex="-1"></a>pal <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;purple&quot;</span>,<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;cyan&quot;</span>,<span class="st">&quot;green&quot;</span>,<span class="st">&quot;yellow&quot;</span>,<span class="st">&quot;red&quot;</span>))</span>
<span id="cb274-2"><a href="ch-03.html#cb274-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(raster_chelsa_crop, <span class="at">col =</span> <span class="fu">pal</span>(<span class="dv">20</span>), <span class="at">xlab=</span><span class="st">&quot;longitude&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;latitude&quot;</span>)</span>
<span id="cb274-3"><a href="ch-03.html#cb274-3" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(lon_lae, lat_lae, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-163-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>We have seen some of the basic functions in the package <code>raster</code>. Now we will load one more package to work with raster files. We use the package <code>rasterVis</code>. This package contains methods for enhanced visualization and interaction with raster data. It implements visualization methods for quantitative data and categorical data, both for univariate and multivariate rasters. It also provides methods to display spatiotemporal rasters, and vector fields.</p>
<div class="sourceCode" id="cb275"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb275-1"><a href="ch-03.html#cb275-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rasterVis)</span>
<span id="cb275-2"><a href="ch-03.html#cb275-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb275-3"><a href="ch-03.html#cb275-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb275-4"><a href="ch-03.html#cb275-4" aria-hidden="true" tabindex="-1"></a>mapTheme <span class="ot">&lt;-</span> <span class="fu">rasterTheme</span>(<span class="at">region =</span> <span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">&quot;RdYlBu&quot;</span>)))</span>
<span id="cb275-5"><a href="ch-03.html#cb275-5" aria-hidden="true" tabindex="-1"></a>plt <span class="ot">&lt;-</span> <span class="fu">levelplot</span>(raster_chelsa_crop, <span class="at">margin=</span><span class="cn">FALSE</span>, <span class="at">par.settings =</span> mapTheme)</span>
<span id="cb275-6"><a href="ch-03.html#cb275-6" aria-hidden="true" tabindex="-1"></a>plt</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-164-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p><strong>Checkpoint</strong></p>
<p>Create your own small raster with 100 grid cells. Give it values, a projection (e.g. ‘+proj=utm +zone=48 +datum=WGS84’) and plot it.</p>
<p><em>Bonus</em>: If you chose the example projection provided, add <em>europe_shape</em> to your plot.</p>
<p><strong>Solution</strong></p>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="ch-03.html#cb276-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a raster</span></span>
<span id="cb276-2"><a href="ch-03.html#cb276-2" aria-hidden="true" tabindex="-1"></a><span class="co"># short compact version</span></span>
<span id="cb276-3"><a href="ch-03.html#cb276-3" aria-hidden="true" tabindex="-1"></a>myraster <span class="ot">&lt;-</span> <span class="fu">raster</span>(<span class="at">ncol=</span><span class="dv">10</span>, <span class="at">nrow=</span><span class="dv">10</span>)</span>
<span id="cb276-4"><a href="ch-03.html#cb276-4" aria-hidden="true" tabindex="-1"></a><span class="co"># OR </span></span>
<span id="cb276-5"><a href="ch-03.html#cb276-5" aria-hidden="true" tabindex="-1"></a><span class="co"># longer version</span></span>
<span id="cb276-6"><a href="ch-03.html#cb276-6" aria-hidden="true" tabindex="-1"></a>myraster <span class="ot">&lt;-</span> <span class="fu">raster</span>()</span>
<span id="cb276-7"><a href="ch-03.html#cb276-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(myraster) <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb276-8"><a href="ch-03.html#cb276-8" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(myraster) <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb276-9"><a href="ch-03.html#cb276-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb276-10"><a href="ch-03.html#cb276-10" aria-hidden="true" tabindex="-1"></a><span class="co"># add values to the raster, this can be done in many ways here are two examples:</span></span>
<span id="cb276-11"><a href="ch-03.html#cb276-11" aria-hidden="true" tabindex="-1"></a><span class="fu">values</span>(myraster) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncell</span>(myraster)</span>
<span id="cb276-12"><a href="ch-03.html#cb276-12" aria-hidden="true" tabindex="-1"></a><span class="co"># OR </span></span>
<span id="cb276-13"><a href="ch-03.html#cb276-13" aria-hidden="true" tabindex="-1"></a><span class="fu">values</span>(myraster) <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="fu">ncell</span>(myraster))</span>
<span id="cb276-14"><a href="ch-03.html#cb276-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb276-15"><a href="ch-03.html#cb276-15" aria-hidden="true" tabindex="-1"></a><span class="co"># add a projection</span></span>
<span id="cb276-16"><a href="ch-03.html#cb276-16" aria-hidden="true" tabindex="-1"></a><span class="fu">projection</span>(myraster) <span class="ot">&lt;-</span> <span class="st">&quot;+proj=utm +zone=48 +datum=WGS84&quot;</span></span>
<span id="cb276-17"><a href="ch-03.html#cb276-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb276-18"><a href="ch-03.html#cb276-18" aria-hidden="true" tabindex="-1"></a><span class="co"># plot it</span></span>
<span id="cb276-19"><a href="ch-03.html#cb276-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(myraster)</span>
<span id="cb276-20"><a href="ch-03.html#cb276-20" aria-hidden="true" tabindex="-1"></a><span class="co"># for fun let&#39;s see where europe would be on our raster, since it has a projection...</span></span>
<span id="cb276-21"><a href="ch-03.html#cb276-21" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(europe_shape, <span class="at">add=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-165-1.png" width="50%" style="display: block; margin: auto;" /></p>
<div id="aggregating-1" class="section level4" number="3.2.5.1">
<h4><span class="header-section-number">3.2.5.1</span> Aggregating</h4>
<p>A <strong>raster grid</strong> is uniquely defined by an <em>origin</em>, a point that one of the intersections of grid lines, and its resolution, the magnitude of each cell-side. In the R package <code>raster</code>, the origin is defined as the point closest to (0,0) that is still an intersection of grid lines. The functions <code>origin()</code> and <code>res()</code> from the R package <code>raster</code> return the origin and resolution of a raster object. So let’s see what the resolution and origin of the rasters we loaded before are.</p>
<div class="sourceCode" id="cb277"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb277-1"><a href="ch-03.html#cb277-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot; Resolution Raster Landover: &quot;</span>, <span class="fu">res</span>(raster_landcover), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb277-2"><a href="ch-03.html#cb277-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Resolution Raster Chelsa: &quot;</span>, <span class="fu">res</span>(raster_chelsa))</span></code></pre></div>
<pre><code>##  Resolution Raster Landover:  0.002777778 0.002777778 
##  Resolution Raster Chelsa:  0.008333333 0.008333333</code></pre>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="ch-03.html#cb279-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot; Origin Raster Landover: &quot;</span>, <span class="fu">origin</span>(raster_landcover), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb279-2"><a href="ch-03.html#cb279-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Origin Raster Chelsa: &quot;</span>, <span class="fu">origin</span>(raster_chelsa))</span></code></pre></div>
<pre><code>##  Origin Raster Landover:  0.001388889 -0.001388889 
##  Origin Raster Chelsa:  -0.000139609 -0.000139249</code></pre>
<p>Landcover has a much higher resolution. Let’s re-grid the Landcover raster to match the grid of the Chelsa raster. Aggregating means resampling an input raster to a coarser resolution based on a specified aggregation strategy. We now determine the aggregation factor by dividing the resolution of the landcover raster by that of the Chelsa raster. We do it both for the longitude and latitude to make sure they match.</p>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="ch-03.html#cb281-1" aria-hidden="true" tabindex="-1"></a>res_landcover <span class="ot">&lt;-</span> <span class="fu">res</span>(raster_landcover)</span>
<span id="cb281-2"><a href="ch-03.html#cb281-2" aria-hidden="true" tabindex="-1"></a>res_chelsa <span class="ot">&lt;-</span> <span class="fu">res</span>(raster_chelsa)</span>
<span id="cb281-3"><a href="ch-03.html#cb281-3" aria-hidden="true" tabindex="-1"></a>factor_agg_lon <span class="ot">&lt;-</span> res_chelsa[<span class="dv">1</span>] <span class="sc">/</span> res_landcover[<span class="dv">1</span>]</span>
<span id="cb281-4"><a href="ch-03.html#cb281-4" aria-hidden="true" tabindex="-1"></a>factor_agg_lat <span class="ot">&lt;-</span> res_chelsa[<span class="dv">2</span>] <span class="sc">/</span> res_landcover[<span class="dv">2</span>]</span>
<span id="cb281-5"><a href="ch-03.html#cb281-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb281-6"><a href="ch-03.html#cb281-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb281-7"><a href="ch-03.html#cb281-7" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot; Longitudinal Aggregation Factor: &quot;</span>, factor_agg_lon, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb281-8"><a href="ch-03.html#cb281-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Latitudinal Aggregation Factor: &quot;</span>, factor_agg_lon)</span></code></pre></div>
<pre><code>##  Longitudinal Aggregation Factor:  3 
##  Latitudinal Aggregation Factor:  3</code></pre>
<p>First, we observed that the land cover raster has a higher resolution than the temperature raster, now we now it is 3 times higher. So we need to transform the landcover raster to make sure it has the same grid (same <em>origin</em> and <em>resolution</em>) as the temperature raster. This will allow us to combine them in a potential analysis. We achieve this using <code>aggregate()</code> from the R package <code>raster</code> to obtain a sum of edges raster at 1000m x 1000m resolution.</p>
<p>The aggregate function takes three main arguments.
* <strong><em>x</em></strong> : the raster object to aggregate,
* <strong><em>fact</em></strong> : aggregation factor, i.e. the number of current cells that will make up the side of one of the new cells, and,
* <strong><em>fun</em></strong> : the function to apply to summarize the raster values of the fact^2 (in this case 3ˆ2=9) cells.</p>
<p>We aggregate the landcover raster at 1km using the aggregation factor calculated above.</p>
<div class="sourceCode" id="cb283"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb283-1"><a href="ch-03.html#cb283-1" aria-hidden="true" tabindex="-1"></a>raster_landcover_crop_agg <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(raster_landcover_crop, <span class="at">fact =</span> factor_agg_lon, <span class="at">fun =</span> modal)</span></code></pre></div>
<p>We check the resolutions and origin of the two rasters and plot our the new landcover one.</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="ch-03.html#cb284-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot; Resolution Landcover: &quot;</span>, <span class="fu">res</span>(raster_landcover_crop_agg), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb284-2"><a href="ch-03.html#cb284-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Resolution Chelsa:    &quot;</span>, <span class="fu">res</span>(raster_chelsa), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb284-3"><a href="ch-03.html#cb284-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Origin Landcover:    &quot;</span>, <span class="fu">origin</span>(raster_landcover_crop_agg), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb284-4"><a href="ch-03.html#cb284-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Origin Chelsa:       &quot;</span>, <span class="fu">origin</span>(raster_chelsa), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>##  Resolution Landcover:  0.008333333 0.008333333 
##  Resolution Chelsa:     0.008333333 0.008333333 
##  Origin Landcover:     -0.001388889 -0.004166667 
##  Origin Chelsa:        -0.000139609 -0.000139249</code></pre>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="ch-03.html#cb286-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(raster_landcover_crop_agg, <span class="at">legend=</span><span class="cn">FALSE</span>, <span class="at">xlab=</span><span class="st">&quot;longitude&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;latitude&quot;</span>)</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-170-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>We can see that the resolutions are now the same but the origin still differs. To get matching origins we must do another step: align the rasters.</p>
</div>
<div id="aligning" class="section level4" number="3.2.5.2">
<h4><span class="header-section-number">3.2.5.2</span> Aligning</h4>
<p>Aligning rasters is done using the function <code>resample()</code>. There are two main resampling methods: <em>Nearest Neighbour</em> or <em>Bilinear Interpolation</em>, which method is used depends upon the input data and its use after the operation is performed.</p>
<p><strong><em>Nearest Neighbour</em></strong> is best used for categorical data like land-use classification or slope classification. The values that go into the grid stay exactly the same, a 2 comes out as a 2, and 99 comes out as 99. The value of the output cell is determined by the nearest cell center on the input grid. Nearest Neighbor can be used on continuous data but the results can be blocky.</p>
<p><strong><em>Bilinear Interpolation</em></strong> uses a weighted average of the four nearest cell centers. The closer an input cell center is to the output cell center, the higher the influence of its value is on the output cell value. This means that the output value could be different than the nearest input, but is always within the same range of values as the input. Since the values can change, Bilinear is not recommended for categorical data. Instead, it should be used for continuous data like elevation and raw slope values.</p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="ch-03.html#cb287-1" aria-hidden="true" tabindex="-1"></a>raster_landcover_crop_resampl <span class="ot">&lt;-</span> <span class="fu">resample</span>(raster_landcover_crop, raster_chelsa_crop, <span class="at">method=</span><span class="st">&quot;ngb&quot;</span>)</span>
<span id="cb287-2"><a href="ch-03.html#cb287-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(raster_landcover_crop_resampl, <span class="at">legend=</span><span class="cn">FALSE</span>, <span class="at">xlab=</span><span class="st">&quot;longitude&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;latitude&quot;</span>)</span></code></pre></div>
<p><a href="./figures/04_output-crop.png"></a></p>
<p>Now we check if the two rasters are aligned.</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="ch-03.html#cb288-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot; Resolution Landcover: &quot;</span>, <span class="fu">res</span>(raster_landcover_crop_resampl), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb288-2"><a href="ch-03.html#cb288-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Resolution Chelsa:    &quot;</span>, <span class="fu">res</span>(raster_chelsa), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb288-3"><a href="ch-03.html#cb288-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Origin Landcover:   &quot;</span>, <span class="fu">origin</span>(raster_landcover_crop_resampl), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb288-4"><a href="ch-03.html#cb288-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Origin Chelsa:      &quot;</span>, <span class="fu">origin</span>(raster_chelsa), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>##  Resolution Landcover:  0.008333333 0.008333333 
##  Resolution Chelsa:     0.008333333 0.008333333 
##  Origin Landcover:    -0.000139609 -0.000139249 
##  Origin Chelsa:       -0.000139609 -0.000139249</code></pre>
<p>Now the two rasters are aligned.</p>
</div>
<div id="point-extraction" class="section level4" number="3.2.5.3">
<h4><span class="header-section-number">3.2.5.3</span> Point extraction</h4>
<p>In this section, we will show you how to extract data from our rasters or spatial data. We will be extracting the values for the locations of our towers. The function to extract values is aptly named <code>extract()</code>. First, we specify the raster from which the values should be extracted and then the exact sites at which the values should be extracted. Here, we combine all this into a neat dataframe.</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="ch-03.html#cb290-1" aria-hidden="true" tabindex="-1"></a>df_sites_temp <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;./data/fluxnet_site_info_reduced.csv&quot;</span>)</span>
<span id="cb290-2"><a href="ch-03.html#cb290-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb290-3"><a href="ch-03.html#cb290-3" aria-hidden="true" tabindex="-1"></a>df_sites_temp <span class="ot">&lt;-</span> <span class="fu">extract</span>(raster_chelsa, sp_sites, <span class="at">sp =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb290-4"><a href="ch-03.html#cb290-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb290-5"><a href="ch-03.html#cb290-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">right_join</span>(df_sites_temp, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;lon&quot;</span>, <span class="st">&quot;lat&quot;</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb290-6"><a href="ch-03.html#cb290-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">temp_chelsa =</span> Chelsa_t_mean_2006<span class="fl">.2012</span>)</span></code></pre></div>
<p>In the lectures, you have already, heard that we must always be aware of modelled data and be aware of its limitations. Therefore, we don’t just want to use CHELSA temperature data. It is important to consider different models and compare them. Hence, we’ll look at another climate model. We load and plot the provided rasters showing the mean annual temperatures for the Worldclim data for the years 2006-2012. WorldClim is a set of global climate layers (gridded climate data) with a spatial resolution of about 1 km2. We extract values from Worldclim at the same tower sites and add them to the <em>df_sites</em> data frame. Now we have a tidy dataframe with the site, latitude and longitude, temperatures for Chelsa and WorldClim at each tower location.</p>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="ch-03.html#cb291-1" aria-hidden="true" tabindex="-1"></a>raster_worldclim <span class="ot">&lt;-</span> <span class="fu">raster</span>(<span class="st">&quot;./data/WC_mean_t_2006-2012.tif&quot;</span>)</span>
<span id="cb291-2"><a href="ch-03.html#cb291-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb291-3"><a href="ch-03.html#cb291-3" aria-hidden="true" tabindex="-1"></a>df_sites_temp <span class="ot">&lt;-</span> <span class="fu">extract</span>(raster_worldclim, sp_sites, <span class="at">sp =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb291-4"><a href="ch-03.html#cb291-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb291-5"><a href="ch-03.html#cb291-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">right_join</span>(df_sites_temp, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;lon&quot;</span>, <span class="st">&quot;lat&quot;</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb291-6"><a href="ch-03.html#cb291-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">temp_wc =</span> WC_mean_t_2006<span class="fl">.2012</span>)</span>
<span id="cb291-7"><a href="ch-03.html#cb291-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb291-8"><a href="ch-03.html#cb291-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_sites_temp)</span></code></pre></div>
<pre><code>## # A tibble: 6 x 5
##   temp_wc   lon   lat temp_chelsa site  
##     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt; 
## 1  -0.633  4.52  51.3        2750 BE-Bra
## 2  -1.45   4.75  50.6        2741 BE-Lon
## 3  -3.75   6.00  50.3        2716 BE-Vie
## 4  -4.76   8.41  47.2        2730 CH-Cha
## 5  -4.97   8.54  47.1        2699 CH-Fru
## 6  -5.38   8.36  47.5        2710 CH-Lae</code></pre>
<p>We get the monthly mean temperature of Jan 2006 from FLUXNET data.</p>
<div class="sourceCode" id="cb293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb293-1"><a href="ch-03.html#cb293-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lubridate)</span>
<span id="cb293-2"><a href="ch-03.html#cb293-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;./data/ddf_allsites_nested_joined.RData&quot;</span>)</span>
<span id="cb293-3"><a href="ch-03.html#cb293-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb293-4"><a href="ch-03.html#cb293-4" aria-hidden="true" tabindex="-1"></a>df_sites_temp <span class="ot">&lt;-</span> ddf_allsites_nested_joined <span class="sc">%&gt;%</span> </span>
<span id="cb293-5"><a href="ch-03.html#cb293-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">site =</span> siteid, data) <span class="sc">%&gt;%</span> </span>
<span id="cb293-6"><a href="ch-03.html#cb293-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(data) <span class="sc">%&gt;%</span> </span>
<span id="cb293-7"><a href="ch-03.html#cb293-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(site, TIMESTAMP, TA_F) <span class="sc">%&gt;%</span> </span>
<span id="cb293-8"><a href="ch-03.html#cb293-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">month =</span> <span class="fu">month</span>(TIMESTAMP), <span class="at">year =</span> <span class="fu">year</span>(TIMESTAMP)) <span class="sc">%&gt;%</span> </span>
<span id="cb293-9"><a href="ch-03.html#cb293-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(month <span class="sc">==</span> <span class="dv">1</span>, year <span class="sc">==</span> <span class="dv">2006</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb293-10"><a href="ch-03.html#cb293-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(site) <span class="sc">%&gt;%</span> </span>
<span id="cb293-11"><a href="ch-03.html#cb293-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">temp_fluxnet =</span> <span class="fu">mean</span>(TA_F)) <span class="sc">%&gt;%</span> </span>
<span id="cb293-12"><a href="ch-03.html#cb293-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb293-13"><a href="ch-03.html#cb293-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add temp data extracted from spatial files (chelsa and worldclim)</span></span>
<span id="cb293-14"><a href="ch-03.html#cb293-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(df_sites_temp, dplyr<span class="sc">::</span><span class="fu">select</span>(site, temp_wc, temp_chelsa),</span>
<span id="cb293-15"><a href="ch-03.html#cb293-15" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">&quot;site&quot;</span>)</span></code></pre></div>
<p>We compare the climate models by showing the correlations with the tower’s data. The dotted line shows what the linear model would look like if the values of our model and the tower data overlapped. The red line shows the actual linear model of the relationship between the climate model and the tower data. The closer the red line to the dotted line the better the model is at predicting the temperatures measured by the towers.</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="ch-03.html#cb294-1" aria-hidden="true" tabindex="-1"></a>df_sites_temp<span class="sc">$</span>temp_chelsa <span class="ot">&lt;-</span> df_sites_temp<span class="sc">$</span>temp_chelsa<span class="sc">/</span><span class="dv">10</span> <span class="sc">-</span><span class="fl">273.15</span></span>
<span id="cb294-2"><a href="ch-03.html#cb294-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb294-3"><a href="ch-03.html#cb294-3" aria-hidden="true" tabindex="-1"></a><span class="co"># we remove the sites we removed before</span></span>
<span id="cb294-4"><a href="ch-03.html#cb294-4" aria-hidden="true" tabindex="-1"></a>df_sites_temp <span class="ot">&lt;-</span> df_sites_temp[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">16</span>, <span class="dv">34</span>), ]</span>
<span id="cb294-5"><a href="ch-03.html#cb294-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb294-6"><a href="ch-03.html#cb294-6" aria-hidden="true" tabindex="-1"></a>df_sites_temp <span class="sc">%&gt;%</span> </span>
<span id="cb294-7"><a href="ch-03.html#cb294-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> temp_fluxnet, <span class="at">y =</span> temp_chelsa), <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">10</span>,<span class="dv">5</span>)) <span class="sc">+</span></span>
<span id="cb294-8"><a href="ch-03.html#cb294-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb294-9"><a href="ch-03.html#cb294-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept=</span><span class="dv">0</span>, <span class="at">slope=</span><span class="dv">1</span>, <span class="at">linetype=</span><span class="st">&quot;dotted&quot;</span>) <span class="sc">+</span></span>
<span id="cb294-10"><a href="ch-03.html#cb294-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span><span class="st">&#39;lm&#39;</span>, <span class="at">color=</span><span class="st">&quot;red&quot;</span>, <span class="at">size=</span><span class="fl">0.5</span>, <span class="at">se=</span><span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb294-11"><a href="ch-03.html#cb294-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-177-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>We compute the mean squared error (MSE). The MSE calculates the average of the squares of errors of all the data points from a fitted line or model. Put simply this is the difference between the data points (the actual data) and the estimated points given by the line or model. The values are squared to get positive values even if the difference is negative. Smaller values reflect less variation of the data. In a model you want to minimise the MSE, as this reflects that the model’s predicted or estimated points are closer to your actual data.</p>
<div class="sourceCode" id="cb295"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb295-1"><a href="ch-03.html#cb295-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Metrics)</span>
<span id="cb295-2"><a href="ch-03.html#cb295-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mse</span>(df_sites_temp<span class="sc">$</span>temp_fluxnet, df_sites_temp<span class="sc">$</span>temp_chelsa)</span></code></pre></div>
<pre><code>## [1] 2.314917</code></pre>
<div class="sourceCode" id="cb297"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb297-1"><a href="ch-03.html#cb297-1" aria-hidden="true" tabindex="-1"></a>df_sites_temp <span class="sc">%&gt;%</span> </span>
<span id="cb297-2"><a href="ch-03.html#cb297-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> temp_fluxnet, <span class="at">y =</span> temp_wc), <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">10</span>,<span class="dv">5</span>)) <span class="sc">+</span></span>
<span id="cb297-3"><a href="ch-03.html#cb297-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb297-4"><a href="ch-03.html#cb297-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept=</span><span class="dv">0</span>, <span class="at">slope=</span><span class="dv">1</span>, <span class="at">linetype=</span><span class="st">&quot;dotted&quot;</span>) <span class="sc">+</span></span>
<span id="cb297-5"><a href="ch-03.html#cb297-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span><span class="st">&#39;lm&#39;</span>, <span class="at">color=</span><span class="st">&quot;red&quot;</span>, <span class="at">size=</span><span class="fl">0.5</span>, <span class="at">se=</span><span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb297-6"><a href="ch-03.html#cb297-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-179-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>We compute the mean squared error.</p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="ch-03.html#cb298-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mse</span>(df_sites_temp<span class="sc">$</span>temp_fluxnet, df_sites_temp<span class="sc">$</span>temp_wc)</span></code></pre></div>
<pre><code>## [1] 10.43716</code></pre>
<p>We see that Chelsa correlates better (because the mean squared error is smaller) with the temperature values at the towers. We also see that Worldclim compared to CHELSA tends to predict colder temperatures.</p>
<p><strong>Checkpoint</strong></p>
<p>Above, we compared the CHELSA temperatures with the tower temperatures. Now we want to visually see which of the two data had the higher mean for each tower location.
Plot the map of europe and add the tower locations as points. Colour the points so that for a higher CHESLA mean they are one colour and a higher tower temperature another.</p>
<p><strong>Solution</strong></p>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="ch-03.html#cb300-1" aria-hidden="true" tabindex="-1"></a><span class="co"># There are more way to do this but we used the function ifelse(). </span></span>
<span id="cb300-2"><a href="ch-03.html#cb300-2" aria-hidden="true" tabindex="-1"></a><span class="co"># It takes 3 components: a condition, what to do if the condition applies and what to do if it doesn&#39;t apply.</span></span>
<span id="cb300-3"><a href="ch-03.html#cb300-3" aria-hidden="true" tabindex="-1"></a><span class="co"># So here if chelsa temperatures are higher we want the point to be purple otherwise orange:</span></span>
<span id="cb300-4"><a href="ch-03.html#cb300-4" aria-hidden="true" tabindex="-1"></a>cols_chelsa_flux <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(df_sites_temp<span class="sc">$</span>temp_chelsa <span class="sc">&gt;</span> df_sites_temp<span class="sc">$</span>temp_fluxnet,<span class="st">&quot;purple&quot;</span>,<span class="st">&quot;orange&quot;</span>)</span>
<span id="cb300-5"><a href="ch-03.html#cb300-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb300-6"><a href="ch-03.html#cb300-6" aria-hidden="true" tabindex="-1"></a><span class="co"># We could also do different shapes if we wanted...</span></span>
<span id="cb300-7"><a href="ch-03.html#cb300-7" aria-hidden="true" tabindex="-1"></a>pch_chelsa_flux <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(df_sites_temp<span class="sc">$</span>temp_chelsa <span class="sc">&gt;</span> df_sites_temp<span class="sc">$</span>temp_fluxnet, <span class="dv">19</span> ,<span class="dv">17</span>)</span>
<span id="cb300-8"><a href="ch-03.html#cb300-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb300-9"><a href="ch-03.html#cb300-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot the result using base R:</span></span>
<span id="cb300-10"><a href="ch-03.html#cb300-10" aria-hidden="true" tabindex="-1"></a><span class="co"># plot(europe_shape)</span></span>
<span id="cb300-11"><a href="ch-03.html#cb300-11" aria-hidden="true" tabindex="-1"></a><span class="co"># points(df_sites_temp$lon,df_sites_temp$lat, pch=pch_diff, col = col_checkpoint)</span></span>
<span id="cb300-12"><a href="ch-03.html#cb300-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb300-13"><a href="ch-03.html#cb300-13" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot the result using ggplot:</span></span>
<span id="cb300-14"><a href="ch-03.html#cb300-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb300-15"><a href="ch-03.html#cb300-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_polygon</span>(<span class="at">data =</span> europe_shape, <span class="fu">aes</span>(<span class="at">x =</span> long, <span class="at">y =</span> lat, <span class="at">group =</span> group), <span class="at">fill=</span><span class="cn">NA</span>, <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb300-16"><a href="ch-03.html#cb300-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> df_sites_temp, <span class="fu">aes</span>(<span class="at">x =</span> lon, <span class="at">y =</span> lat), <span class="at">color =</span> cols_chelsa_flux, <span class="at">shape =</span> pch_chelsa_flux, <span class="at">size=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb300-17"><a href="ch-03.html#cb300-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Longitude&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Latitude&quot;</span>) <span class="sc">+</span></span>
<span id="cb300-18"><a href="ch-03.html#cb300-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()<span class="sc">+</span> </span>
<span id="cb300-19"><a href="ch-03.html#cb300-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_quickmap</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-181-1.png" width="50%" style="display: block; margin: auto;" /></p>
</div>
<div id="correlations-with-gpp" class="section level4" number="3.2.5.4">
<h4><span class="header-section-number">3.2.5.4</span> Correlations with GPP</h4>
<p>We already had a look at GPP in Chapter <a href="ch-02.html#ch-02">2</a>. In this section we want to correlate Chelsa temperature and landcover with GPP.</p>
<p>In models we often take multiple variables and correlate them to another variable (here GPP). By seeing if they correlate well we can see if any of the variables make good so-called <em>predictors</em>. Let’s see how well landcover and temperature correlate with GPP.</p>
<p>We start off by extracting the mean GPP at the tower sites. To do this we group our data by the <em>siteid</em> (or tower) and the year. This way we can get the GPP across a year and then an average across all years for each site.</p>
<div class="sourceCode" id="cb301"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb301-1"><a href="ch-03.html#cb301-1" aria-hidden="true" tabindex="-1"></a>GPP <span class="ot">&lt;-</span> ddf_allsites_nested_joined <span class="sc">%&gt;%</span> </span>
<span id="cb301-2"><a href="ch-03.html#cb301-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(data) <span class="sc">%&gt;%</span> </span>
<span id="cb301-3"><a href="ch-03.html#cb301-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">year =</span> <span class="fu">year</span>(TIMESTAMP)) <span class="sc">%&gt;%</span> </span>
<span id="cb301-4"><a href="ch-03.html#cb301-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(siteid, year) <span class="sc">%&gt;%</span> </span>
<span id="cb301-5"><a href="ch-03.html#cb301-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">gpp_ann =</span> <span class="fu">sum</span>(GPP_NT_VUT_REF)) <span class="sc">%&gt;%</span> </span>
<span id="cb301-6"><a href="ch-03.html#cb301-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb301-7"><a href="ch-03.html#cb301-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(siteid) <span class="sc">%&gt;%</span> </span>
<span id="cb301-8"><a href="ch-03.html#cb301-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">gpp_meanann =</span> <span class="fu">mean</span>(gpp_ann)) <span class="sc">%&gt;%</span> </span>
<span id="cb301-9"><a href="ch-03.html#cb301-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(siteid, gpp_meanann)</span>
<span id="cb301-10"><a href="ch-03.html#cb301-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb301-11"><a href="ch-03.html#cb301-11" aria-hidden="true" tabindex="-1"></a><span class="co"># we remove the data we don&#39;t have in the other datasets for the correlations</span></span>
<span id="cb301-12"><a href="ch-03.html#cb301-12" aria-hidden="true" tabindex="-1"></a>GPP <span class="ot">&lt;-</span> GPP[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">14</span>, <span class="dv">16</span>, <span class="dv">34</span>), ]</span></code></pre></div>
<p>We then extract landcover data at the tower sites.</p>
<div class="sourceCode" id="cb302"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb302-1"><a href="ch-03.html#cb302-1" aria-hidden="true" tabindex="-1"></a>df_sites_landcover <span class="ot">&lt;-</span> <span class="fu">extract</span>(raster_landcover, sp_sites, <span class="at">sp =</span> <span class="cn">TRUE</span>)</span>
<span id="cb302-2"><a href="ch-03.html#cb302-2" aria-hidden="true" tabindex="-1"></a>df_sites_landcover <span class="ot">&lt;-</span> <span class="fu">as.tibble</span>(df_sites_landcover)</span>
<span id="cb302-3"><a href="ch-03.html#cb302-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb302-4"><a href="ch-03.html#cb302-4" aria-hidden="true" tabindex="-1"></a><span class="co"># we remove the sites we removed before</span></span>
<span id="cb302-5"><a href="ch-03.html#cb302-5" aria-hidden="true" tabindex="-1"></a>df_sites_landcover <span class="ot">&lt;-</span> df_sites_landcover[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">16</span>, <span class="dv">34</span>), ]</span></code></pre></div>
<p>And create a dataframe with all the extracted data from the tower sites to then make correlations.</p>
<div class="sourceCode" id="cb303"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb303-1"><a href="ch-03.html#cb303-1" aria-hidden="true" tabindex="-1"></a>landcover <span class="ot">&lt;-</span> df_sites_landcover<span class="sc">$</span>Globcover_EU</span>
<span id="cb303-2"><a href="ch-03.html#cb303-2" aria-hidden="true" tabindex="-1"></a>chelsa <span class="ot">&lt;-</span> df_sites_temp<span class="sc">$</span>temp_chelsa</span>
<span id="cb303-3"><a href="ch-03.html#cb303-3" aria-hidden="true" tabindex="-1"></a>df_corr <span class="ot">&lt;-</span> <span class="fu">cbind</span>(landcover, chelsa, GPP)</span>
<span id="cb303-4"><a href="ch-03.html#cb303-4" aria-hidden="true" tabindex="-1"></a>df_corr <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(df_corr)</span>
<span id="cb303-5"><a href="ch-03.html#cb303-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_corr)</span></code></pre></div>
<pre><code>##   landcover chelsa siteid gpp_meanann
## 1        50   1.85 BE-Bra          NA
## 2        20   0.95 BE-Lon    1348.667
## 3        90  -1.55 BE-Vie    1792.497
## 4       120  -0.15 CH-Cha    2478.429
## 5       140  -6.95 CH-Dav    1166.543
## 6        50  -3.25 CH-Fru    2106.319</code></pre>
<p>Though the landcover column contains numbers, it is actually categorical data. Each number represents a specific landcover type, so we add a column with the corresponding the landcover category name.</p>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="ch-03.html#cb305-1" aria-hidden="true" tabindex="-1"></a>df_corr <span class="ot">&lt;-</span> df_corr <span class="sc">%&gt;%</span></span>
<span id="cb305-2"><a href="ch-03.html#cb305-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">landcover_cat =</span> <span class="fu">case_when</span>(landcover <span class="sc">==</span> <span class="dv">14</span> <span class="sc">~</span> <span class="st">&quot;cropland&quot;</span>,</span>
<span id="cb305-3"><a href="ch-03.html#cb305-3" aria-hidden="true" tabindex="-1"></a>        landcover <span class="sc">==</span> <span class="dv">20</span> <span class="sc">~</span> <span class="st">&quot;mosaic cropland&quot;</span>,</span>
<span id="cb305-4"><a href="ch-03.html#cb305-4" aria-hidden="true" tabindex="-1"></a>        landcover <span class="sc">==</span> <span class="dv">50</span> <span class="sc">~</span> <span class="st">&quot;deciduous forest&quot;</span>,</span>
<span id="cb305-5"><a href="ch-03.html#cb305-5" aria-hidden="true" tabindex="-1"></a>        landcover <span class="sc">==</span> <span class="dv">70</span> <span class="sc">~</span> <span class="st">&quot;evergreen forest&quot;</span>,</span>
<span id="cb305-6"><a href="ch-03.html#cb305-6" aria-hidden="true" tabindex="-1"></a>        landcover <span class="sc">==</span> <span class="dv">90</span> <span class="sc">~</span> <span class="st">&quot;needleleaved forest&quot;</span>,</span>
<span id="cb305-7"><a href="ch-03.html#cb305-7" aria-hidden="true" tabindex="-1"></a>        landcover <span class="sc">==</span> <span class="dv">110</span> <span class="sc">~</span> <span class="st">&quot;mosaic forest&quot;</span>,</span>
<span id="cb305-8"><a href="ch-03.html#cb305-8" aria-hidden="true" tabindex="-1"></a>        landcover <span class="sc">==</span> <span class="dv">120</span> <span class="sc">~</span> <span class="st">&quot;mosaic forest/shrub/grass&quot;</span>,</span>
<span id="cb305-9"><a href="ch-03.html#cb305-9" aria-hidden="true" tabindex="-1"></a>        landcover <span class="sc">==</span> <span class="dv">130</span> <span class="sc">~</span> <span class="st">&quot;mosaic shrubland&quot;</span>,</span>
<span id="cb305-10"><a href="ch-03.html#cb305-10" aria-hidden="true" tabindex="-1"></a>        landcover <span class="sc">==</span> <span class="dv">140</span> <span class="sc">~</span> <span class="st">&quot;mosaic grassland&quot;</span>,</span>
<span id="cb305-11"><a href="ch-03.html#cb305-11" aria-hidden="true" tabindex="-1"></a>        landcover <span class="sc">==</span> <span class="dv">150</span> <span class="sc">~</span> <span class="st">&quot;sparse vegetation&quot;</span>))</span></code></pre></div>
<p>Back to correlations! Let’s start by correlating the Chelsa temperatures with GPP. We do this by building a simple linear model using the function <code>lm()</code>.</p>
<div class="sourceCode" id="cb306"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb306-1"><a href="ch-03.html#cb306-1" aria-hidden="true" tabindex="-1"></a>reg1 <span class="ot">&lt;-</span> <span class="fu">lm</span>(gpp_meanann<span class="sc">~</span>chelsa, <span class="at">data =</span> df_corr) </span></code></pre></div>
<p>To visualise the correlation we plot our temperature data with the linear model.</p>
<div class="sourceCode" id="cb307"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb307-1"><a href="ch-03.html#cb307-1" aria-hidden="true" tabindex="-1"></a>df_corr <span class="sc">%&gt;%</span> </span>
<span id="cb307-2"><a href="ch-03.html#cb307-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> chelsa, <span class="at">y =</span> gpp_meanann)) <span class="sc">+</span></span>
<span id="cb307-3"><a href="ch-03.html#cb307-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb307-4"><a href="ch-03.html#cb307-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span><span class="st">&#39;lm&#39;</span>, <span class="at">color=</span><span class="st">&quot;red&quot;</span>, <span class="at">size=</span><span class="fl">0.5</span>, <span class="at">se=</span><span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb307-5"><a href="ch-03.html#cb307-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Chelsa temperature&quot;</span>) <span class="sc">+</span></span>
<span id="cb307-6"><a href="ch-03.html#cb307-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;GPP mean [µmolCO2 m-2 s-1]&quot;</span>) <span class="sc">+</span></span>
<span id="cb307-7"><a href="ch-03.html#cb307-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-187-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>Then we make a linear model for GPP and landcover.</p>
<div class="sourceCode" id="cb308"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb308-1"><a href="ch-03.html#cb308-1" aria-hidden="true" tabindex="-1"></a>reg2 <span class="ot">&lt;-</span> <span class="fu">lm</span>(gpp_meanann<span class="sc">~</span>landcover, <span class="at">data=</span>df_corr)</span></code></pre></div>
<p>… and again plot it. However, since landcover is categorical data we make a boxplot instead. We’ve also added a general trendline but remember this could be quite different if we correclated data from a specific landcover category with GPP.</p>
<div class="sourceCode" id="cb309"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb309-1"><a href="ch-03.html#cb309-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>df_corr, <span class="fu">aes</span>(<span class="at">x=</span>landcover_cat, <span class="at">y=</span>gpp_meanann, <span class="at">group=</span><span class="fu">as.factor</span>(landcover_cat)))<span class="sc">+</span> </span>
<span id="cb309-2"><a href="ch-03.html#cb309-2" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb309-3"><a href="ch-03.html#cb309-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Landcover Category&quot;</span>, <span class="at">y =</span> <span class="st">&quot;GPP mean [µmolCO2 m-2 s-1]&quot;</span>) <span class="sc">+</span></span>
<span id="cb309-4"><a href="ch-03.html#cb309-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb309-5"><a href="ch-03.html#cb309-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">size=</span><span class="dv">12</span>))</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-189-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>We don’t see a particularly strong correlation, thus we need to consider other predictors and more complex models. You will discuss that later in the class.</p>
</div>
</div>
<div id="key-points-of-the-tutorial" class="section level3" number="3.2.6">
<h3><span class="header-section-number">3.2.6</span> Key points of the tutorial</h3>
<p>Integrating remote sensing data: we extract NDVI in the area around our eddy flux towers from the MODISTools package and conver it into a raster format.</p>
<ul>
<li>Collapse NDVI measured within an area around the eddy flux towers to calculate the mean across years</li>
<li>Merge this temporal data with our eddy flux data</li>
<li>Apply linear regression to test whether the addition of NDVI data to our model of GPP and PPFD_IN improves its accuracy</li>
</ul>
<p>Points:
the eddy flux tower sites are plotted as points on a map of Europe.</p>
<ul>
<li>Extract coordinates of the tower sites from the metadata</li>
<li>Transform the points into SpatialPoints with the sp package and plot them on a map of Europe</li>
</ul>
<p>Shapefiles: we add country and border information to our tower site plots.</p>
<ul>
<li>Extract the information about the spatial polygons in the europe_shape shapefile and apply one coordinate reference system</li>
<li>Add the SpatialPoints matrix and shapefile data europe_shape to the tower site dataframe</li>
<li>Filter out the countries with towers and visualize only them</li>
</ul>
<p>Rasters: we investigate whether the variables landcover and temperature can explain GPP.</p>
<ul>
<li>Load landcover data from Globcover temperature data from CHELSA in the form of rasters and plot them around our tower sites with the raster and rasterVis packages</li>
<li>Apply the same resolution (aggregate) and origin (align) to the climate and landcover rasters</li>
<li>Compare modeled climate data from CHELSA and Worldclim to our fluxnet data by extracting the temperatures at the tower sites and calculating the MSE</li>
<li>Finally, landcover and CHELSA modeled temperatures are plugged into a linear regression to see if they are significantly correlated with GPP</li>
</ul>
<p>Bonus: here we incorporate plant species data from GBIF to determine their effect on GPP with the help of PCA.</p>
<ul>
<li>Create a presence-absence matrix for all locations and species</li>
<li>Check species data for possible sampling biases by visualizing the number of occurrence records</li>
<li>Plot species richness on the map</li>
<li>Add species traits to our dataset with a for loop</li>
<li>Use PCA to determine which traits explain the most variation in GPP</li>
</ul>
</div>
<div id="bonus-species-occurrence-trait-data-and-pcas" class="section level3" number="3.2.7">
<h3><span class="header-section-number">3.2.7</span> Bonus: Species Occurrence, Trait Data and PCAs</h3>
<p>In functional ecology, we focus on the function that a species has in a community. This sub-discipline of ecology represents the intersection between ecological patterns and the mechanisms that underlie them and focuses on traits represented in large number of species. Here, we use plant distribution data around the fluxnet towers to investigate the plant community composition for each location. The community composition will allow us to get insights on which plant traits that are dominant. This is relevant because plant traits are both responsive to local climate and strong predictors of primary productivity. For example, Specific Leaf Area (SLA) and plant height, are linked to GPP and affected by local climate. Functional traits capture core differences in the strategies plants use to acquire and invest resources. Most woody plants have the same basic physiological function and key resource requirements, however species differ considerably in the rates at which resources are acquired, invested into different tissues, and lost via turnover. Traits for example moderate plant responses to light environment and determine shade tolerance. They have been shown to affect growth depending on plant size. In the first step, we use distribution data from the Global Biodiversity Information Facility (GBIF) to obtain the community composition in each location.</p>
<p>GBIF is an international organization that makes biodiversity data publicly available. The data are provided by many institutions from around the world and GBIF gathers these data and makes them accessible and searchable through their platform. Data available on GBIF are primarily distribution records on plants, animals, fungi, and microbes for the world, and scientific names data. This is very useful information for ecological analysis. With the “rgbif” package we can download occurrence data directly from the GBIF database to R.</p>
<div id="species-occurrences" class="section level5" number="3.2.7.0.1">
<h5><span class="header-section-number">3.2.7.0.1</span> Species Occurrences</h5>
<p>We start by to loading the relevant packages.</p>
<div class="sourceCode" id="cb310"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb310-1"><a href="ch-03.html#cb310-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rgbif)</span>
<span id="cb310-2"><a href="ch-03.html#cb310-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rgeos)</span></code></pre></div>
<p>In this first, section we will give you an example of how to download data from gbif for the CH-Lae site.</p>
<div class="sourceCode" id="cb311"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb311-1"><a href="ch-03.html#cb311-1" aria-hidden="true" tabindex="-1"></a><span class="co"># combine the CH-Lae latitude and longitude</span></span>
<span id="cb311-2"><a href="ch-03.html#cb311-2" aria-hidden="true" tabindex="-1"></a>CHLae_lonlat <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cbind</span>(<span class="at">longitude=</span>lon_lae, <span class="at">latitude=</span>lat_lae))</span>
<span id="cb311-3"><a href="ch-03.html#cb311-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb311-4"><a href="ch-03.html#cb311-4" aria-hidden="true" tabindex="-1"></a><span class="co"># make them into &#39;SpatialPoints&#39;, this is an alternative method to the one used above</span></span>
<span id="cb311-5"><a href="ch-03.html#cb311-5" aria-hidden="true" tabindex="-1"></a><span class="fu">coordinates</span>(CHLae_lonlat) <span class="ot">&lt;-</span> <span class="er">~</span>longitude <span class="sc">+</span> latitude</span>
<span id="cb311-6"><a href="ch-03.html#cb311-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb311-7"><a href="ch-03.html#cb311-7" aria-hidden="true" tabindex="-1"></a><span class="co"># assign a projection and coordinate reference system</span></span>
<span id="cb311-8"><a href="ch-03.html#cb311-8" aria-hidden="true" tabindex="-1"></a><span class="fu">crs</span>(CHLae_lonlat) <span class="ot">&lt;-</span> <span class="st">&quot;+init=epsg:4326 +proj=longlat +datum=WGS84 +no_defs&quot;</span></span>
<span id="cb311-9"><a href="ch-03.html#cb311-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb311-10"><a href="ch-03.html#cb311-10" aria-hidden="true" tabindex="-1"></a><span class="co"># create a roughly 5.5km buffer around the CH-Lae tower</span></span>
<span id="cb311-11"><a href="ch-03.html#cb311-11" aria-hidden="true" tabindex="-1"></a>TowerBuffer <span class="ot">&lt;-</span> <span class="fu">gBuffer</span>(CHLae_lonlat, <span class="at">width=</span><span class="fl">0.05</span>)</span>
<span id="cb311-12"><a href="ch-03.html#cb311-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb311-13"><a href="ch-03.html#cb311-13" aria-hidden="true" tabindex="-1"></a><span class="co"># transform the tower buffer into polygons</span></span>
<span id="cb311-14"><a href="ch-03.html#cb311-14" aria-hidden="true" tabindex="-1"></a>tower.polygon <span class="ot">&lt;-</span> <span class="fu">polygons</span>(TowerBuffer)</span></code></pre></div>
<p>Then transform the polygon to wkt format which is needed for the gbif input:</p>
<div class="sourceCode" id="cb312"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb312-1"><a href="ch-03.html#cb312-1" aria-hidden="true" tabindex="-1"></a>tower.polygon.wkt <span class="ot">&lt;-</span> <span class="fu">writeWKT</span>(tower.polygon)</span></code></pre></div>
<p>Now we are ready to download occurrence records of plants around the CH-Lae tower from gbif with <code>occ_search()</code> function.
We add the hasCoordinate = T, to make sure we only download data with coordinates.
The <em>geometry</em> argument will be equal to our constructed polygon of the buffer.</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="ch-03.html#cb313-1" aria-hidden="true" tabindex="-1"></a>occ_CHlae <span class="ot">&lt;-</span> <span class="fu">occ_data</span>(<span class="at">taxonKey =</span> <span class="dv">7707728</span>, <span class="at">hasCoordinate =</span> T, </span>
<span id="cb313-2"><a href="ch-03.html#cb313-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">geometry =</span> tower.polygon.wkt, <span class="at">limit =</span> <span class="dv">50000</span>)</span>
<span id="cb313-3"><a href="ch-03.html#cb313-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb313-4"><a href="ch-03.html#cb313-4" aria-hidden="true" tabindex="-1"></a><span class="co"># The output is a list so we extract the actual data since this is of interest</span></span>
<span id="cb313-5"><a href="ch-03.html#cb313-5" aria-hidden="true" tabindex="-1"></a>occ_CHlae <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(occ_CHlae<span class="sc">$</span>data)</span></code></pre></div>
<pre><code>##          key                 scientificName decimalLatitude decimalLongitude
## 1 3028451251 Berberis julianae C.K.Schneid.        47.46482         8.333900
## 2 2646869619            Corylus avellana L.        47.44893         8.356892
## 3 2649395661             Daphne laureola L.        47.48192         8.317558
## 4 2649628860                Hedera helix L.        47.50389         8.346389
## 5 2878758835       Asplenium trichomanes L.        47.46482         8.333900
## 6 2878758840            Thlaspi montanum L.        47.46482         8.333900
##         issues                           datasetKey
## 1              83fdfd3d-3a25-4705-9fbe-3db1d1892b13
## 2 cdround,cudc 14d5676a-2c54-4f94-9023-1e8dcd822aa0
## 3 cdround,cudc 14d5676a-2c54-4f94-9023-1e8dcd822aa0
## 4 cdround,cudc 14d5676a-2c54-4f94-9023-1e8dcd822aa0
## 5              83fdfd3d-3a25-4705-9fbe-3db1d1892b13
## 6              83fdfd3d-3a25-4705-9fbe-3db1d1892b13</code></pre>
<p>Let’s have a quick look at the orders of the species in the polygon around our tower.</p>
<div class="sourceCode" id="cb315"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb315-1"><a href="ch-03.html#cb315-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first we make the maine plot of the points coloured by order with the buffer polygon around them</span></span>
<span id="cb315-2"><a href="ch-03.html#cb315-2" aria-hidden="true" tabindex="-1"></a>plot_sp_occ <span class="ot">&lt;-</span> <span class="fu">ggplot</span>()<span class="sc">+</span></span>
<span id="cb315-3"><a href="ch-03.html#cb315-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>occ_CHlae, <span class="fu">aes</span>(<span class="at">x=</span>decimalLongitude, <span class="at">y=</span>decimalLatitude, <span class="at">color=</span>order))<span class="sc">+</span></span>
<span id="cb315-4"><a href="ch-03.html#cb315-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_polygon</span>(<span class="at">data=</span>tower.polygon, </span>
<span id="cb315-5"><a href="ch-03.html#cb315-5" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x=</span>tower.polygon<span class="sc">@</span>polygons[[<span class="dv">1</span>]]<span class="sc">@</span>Polygons[[<span class="dv">1</span>]]<span class="sc">@</span>coords[,<span class="dv">1</span>], <span class="at">y=</span>tower.polygon<span class="sc">@</span>polygons[[<span class="dv">1</span>]]<span class="sc">@</span>Polygons[[<span class="dv">1</span>]]<span class="sc">@</span>coords[,<span class="dv">2</span>]), <span class="at">fill =</span> <span class="cn">NA</span>, <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>)<span class="sc">+</span></span>
<span id="cb315-6"><a href="ch-03.html#cb315-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span> </span>
<span id="cb315-7"><a href="ch-03.html#cb315-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Longitude&quot;</span>) <span class="sc">+</span> </span>
<span id="cb315-8"><a href="ch-03.html#cb315-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Latitude&quot;</span>) <span class="sc">+</span></span>
<span id="cb315-9"><a href="ch-03.html#cb315-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span>
<span id="cb315-10"><a href="ch-03.html#cb315-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb315-11"><a href="ch-03.html#cb315-11" aria-hidden="true" tabindex="-1"></a><span class="co"># then we add the CH-Lae tower in the middle in red</span></span>
<span id="cb315-12"><a href="ch-03.html#cb315-12" aria-hidden="true" tabindex="-1"></a>plot_sp_occ <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x=</span>lon_lae, <span class="at">y=</span>lat_lae, <span class="at">fill=</span><span class="st">&quot;#FC717F&quot;</span>), <span class="at">size=</span><span class="dv">5</span>, <span class="at">colour=</span><span class="st">&quot;black&quot;</span>, <span class="at">shape=</span><span class="dv">21</span>, <span class="at">stroke=</span><span class="dv">1</span>)</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-195-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>To make a species list and calculate the species richness for that tower we simply extract the unique species</p>
<div class="sourceCode" id="cb316"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb316-1"><a href="ch-03.html#cb316-1" aria-hidden="true" tabindex="-1"></a>sp.list <span class="ot">&lt;-</span> <span class="fu">unique</span>(occ_CHlae<span class="sc">$</span>scientificName)</span>
<span id="cb316-2"><a href="ch-03.html#cb316-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(occ_CHlae<span class="sc">$</span>scientificName))</span></code></pre></div>
<pre><code>## [1] 1254</code></pre>
<p>Normally at this point a species name filtration step would be done. That means the plant names from our species list would be compared to that of ‘The Plant List’ (TPL) using the package <code>Taxonstand</code>.
This is quite time consuming, so we have already done it for the species lists of each tower.
Below is the example code:</p>
<div class="sourceCode" id="cb318"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb318-1"><a href="ch-03.html#cb318-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We retrieve name from taxonstand. To specify how much difference in the names we want to allow we specify diffchar=.</span></span>
<span id="cb318-2"><a href="ch-03.html#cb318-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Taxonstand)</span>
<span id="cb318-3"><a href="ch-03.html#cb318-3" aria-hidden="true" tabindex="-1"></a>name <span class="ot">&lt;-</span> <span class="fu">TPL</span>(sp.list.final, <span class="at">diffchar =</span> <span class="dv">4</span>, <span class="at">drop.lower.level =</span> T)</span>
<span id="cb318-4"><a href="ch-03.html#cb318-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb318-5"><a href="ch-03.html#cb318-5" aria-hidden="true" tabindex="-1"></a><span class="co"># select columns of interest</span></span>
<span id="cb318-6"><a href="ch-03.html#cb318-6" aria-hidden="true" tabindex="-1"></a>name <span class="ot">&lt;-</span> name[, <span class="fu">c</span>(<span class="st">&quot;Taxon&quot;</span>, <span class="st">&quot;New.Genus&quot;</span>, <span class="st">&quot;New.Species&quot;</span>, <span class="st">&quot;New.Taxonomic.status&quot;</span>, <span class="st">&quot;Typo&quot;</span>)]</span>
<span id="cb318-7"><a href="ch-03.html#cb318-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb318-8"><a href="ch-03.html#cb318-8" aria-hidden="true" tabindex="-1"></a><span class="co"># create column with new species name</span></span>
<span id="cb318-9"><a href="ch-03.html#cb318-9" aria-hidden="true" tabindex="-1"></a>name<span class="sc">$</span>autocorrected <span class="ot">&lt;-</span> <span class="fu">paste</span>(name<span class="sc">$</span>New.Genus, name<span class="sc">$</span>New.Species)</span>
<span id="cb318-10"><a href="ch-03.html#cb318-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb318-11"><a href="ch-03.html#cb318-11" aria-hidden="true" tabindex="-1"></a><span class="co"># drop new.species and new.genus column</span></span>
<span id="cb318-12"><a href="ch-03.html#cb318-12" aria-hidden="true" tabindex="-1"></a>name <span class="ot">&lt;-</span> name[, <span class="fu">c</span>(<span class="st">&quot;Taxon&quot;</span>, <span class="st">&quot;autocorrected&quot;</span>, <span class="st">&quot;New.Taxonomic.status&quot;</span>, <span class="st">&quot;Typo&quot;</span>)]</span>
<span id="cb318-13"><a href="ch-03.html#cb318-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb318-14"><a href="ch-03.html#cb318-14" aria-hidden="true" tabindex="-1"></a><span class="co"># change column names</span></span>
<span id="cb318-15"><a href="ch-03.html#cb318-15" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(name) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;submittedname&quot;</span>, <span class="st">&quot;autocorrected&quot;</span>, <span class="st">&quot;taxo_status&quot;</span>, <span class="st">&quot;typo&quot;</span>)</span>
<span id="cb318-16"><a href="ch-03.html#cb318-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb318-17"><a href="ch-03.html#cb318-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Finally, we can drop all species for which we did not find a matching scientific name.</span></span>
<span id="cb318-18"><a href="ch-03.html#cb318-18" aria-hidden="true" tabindex="-1"></a>names.accepted <span class="ot">&lt;-</span> <span class="fu">unique</span>(name[name<span class="sc">$</span>taxo_status<span class="sc">==</span><span class="st">&quot;Accepted&quot;</span>,])</span></code></pre></div>
<div class="sourceCode" id="cb319"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb319-1"><a href="ch-03.html#cb319-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the resolved species names .csv</span></span>
<span id="cb319-2"><a href="ch-03.html#cb319-2" aria-hidden="true" tabindex="-1"></a>species_list_full <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;./data/resolved_species_names.csv&quot;</span>,<span class="at">sep=</span><span class="st">&quot;;&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb320"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb320-1"><a href="ch-03.html#cb320-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check how many species are in the list</span></span>
<span id="cb320-2"><a href="ch-03.html#cb320-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(species_list_full)</span></code></pre></div>
<pre><code>## [1] 7480    1</code></pre>
<p>Next, we can use this information to build a presence-absence matrix for all locations and species, this will allow us to see where each species is found. This can be the basis for many community ecology calculations or species distribution models linking the environmental variables to species’ occurrences to see which are most correlated with certain factors.</p>
<div class="sourceCode" id="cb322"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb322-1"><a href="ch-03.html#cb322-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select all files with resolved in their name</span></span>
<span id="cb322-2"><a href="ch-03.html#cb322-2" aria-hidden="true" tabindex="-1"></a>paths <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="fu">file.path</span>(<span class="st">&quot;./data/&quot;</span>, <span class="st">&quot;species_occurrences&quot;</span>), <span class="at">pattern=</span><span class="st">&#39;resolved_*&#39;</span>, <span class="at">recursive =</span> F, <span class="at">full.names =</span> T)</span>
<span id="cb322-3"><a href="ch-03.html#cb322-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-4"><a href="ch-03.html#cb322-4" aria-hidden="true" tabindex="-1"></a><span class="co"># create a function to read them all in</span></span>
<span id="cb322-5"><a href="ch-03.html#cb322-5" aria-hidden="true" tabindex="-1"></a>read_species <span class="ot">=</span> <span class="cf">function</span>(file_path){</span>
<span id="cb322-6"><a href="ch-03.html#cb322-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">read.csv</span>(file_path,<span class="at">sep=</span><span class="st">&quot;;&quot;</span>))</span>
<span id="cb322-7"><a href="ch-03.html#cb322-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb322-8"><a href="ch-03.html#cb322-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-9"><a href="ch-03.html#cb322-9" aria-hidden="true" tabindex="-1"></a><span class="co"># read all the files in paths in using the function read_species</span></span>
<span id="cb322-10"><a href="ch-03.html#cb322-10" aria-hidden="true" tabindex="-1"></a>species_occurrences <span class="ot">&lt;-</span> <span class="fu">lapply</span>(paths, read_species)</span>
<span id="cb322-11"><a href="ch-03.html#cb322-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-12"><a href="ch-03.html#cb322-12" aria-hidden="true" tabindex="-1"></a><span class="co"># bind them into a df and give an identifier with .id=</span></span>
<span id="cb322-13"><a href="ch-03.html#cb322-13" aria-hidden="true" tabindex="-1"></a>species_occ_df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(species_occurrences, <span class="at">.id=</span><span class="st">&quot;id&quot;</span>)</span>
<span id="cb322-14"><a href="ch-03.html#cb322-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-15"><a href="ch-03.html#cb322-15" aria-hidden="true" tabindex="-1"></a><span class="co"># make a species occurrence (or presence-absence) matrix</span></span>
<span id="cb322-16"><a href="ch-03.html#cb322-16" aria-hidden="true" tabindex="-1"></a>species_occ_mat <span class="ot">&lt;-</span> species_occ_df <span class="sc">%&gt;%</span></span>
<span id="cb322-17"><a href="ch-03.html#cb322-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(id)<span class="sc">%&gt;%</span>                           <span class="co"># group the species lists by the sites</span></span>
<span id="cb322-18"><a href="ch-03.html#cb322-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>() <span class="sc">%&gt;%</span>                            <span class="co"># select only unique species names since some occur multiple times</span></span>
<span id="cb322-19"><a href="ch-03.html#cb322-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">present =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span>                   <span class="co"># create a column where all the species get a 1 (for presence)</span></span>
<span id="cb322-20"><a href="ch-03.html#cb322-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> names.accepted.autocorrected, <span class="at">values_from =</span> present) <span class="sc">%&gt;%</span>   <span class="co"># turn (spread) the column of species names to the column names</span></span>
<span id="cb322-21"><a href="ch-03.html#cb322-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate_all</span>(<span class="fu">funs</span>(<span class="fu">ifelse</span>(<span class="fu">is.na</span>(.), <span class="dv">0</span>, .)))  <span class="co"># replace the missing values with 0</span></span></code></pre></div>
<div class="sourceCode" id="cb323"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb323-1"><a href="ch-03.html#cb323-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(species_occ_mat) <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>) <span class="co"># Only displaying the first 6 featurs of species_occ_mat</span></span></code></pre></div>
<pre><code>## # A tibble: 6 x 6
## # Groups:   id [6]
##   id    `Hedera helix` `Arum italicum` `Asplenium ruta-mura… `Ceterach officina…
##   &lt;chr&gt;          &lt;dbl&gt;           &lt;dbl&gt;                 &lt;dbl&gt;               &lt;dbl&gt;
## 1 1                  1               1                     1                   1
## 2 2                  1               1                     1                   1
## 3 3                  1               1                     1                   1
## 4 4                  1               1                     1                   1
## 5 5                  1               1                     1                   1
## 6 6                  1               1                     1                   1
## # … with 1 more variable: Artemisia vulgaris &lt;dbl&gt;</code></pre>
<p>GBIF is a great source for species occurrences. However, we have to keep in mind that the data collection is biased by regions. For example, we find much fewer plant records in Africa than in Europe. But also within Europe, there are some biases. Therefore we need to be careful with the interpretation.</p>
<div class="sourceCode" id="cb325"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb325-1"><a href="ch-03.html#cb325-1" aria-hidden="true" tabindex="-1"></a>nr_occurrences <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;./data/flx_tower_occurrences.csv&quot;</span>)</span>
<span id="cb325-2"><a href="ch-03.html#cb325-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb325-3"><a href="ch-03.html#cb325-3" aria-hidden="true" tabindex="-1"></a>occurrences <span class="ot">&lt;-</span> nr_occurrences<span class="sc">$</span>nr_occurrences</span></code></pre></div>
<div class="sourceCode" id="cb326"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb326-1"><a href="ch-03.html#cb326-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb326-2"><a href="ch-03.html#cb326-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_polygon</span>(<span class="at">data =</span> europe_shape,</span>
<span id="cb326-3"><a href="ch-03.html#cb326-3" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x =</span> long, <span class="at">y =</span> lat, <span class="at">group =</span> group), <span class="at">fill =</span> <span class="cn">NA</span>, <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb326-4"><a href="ch-03.html#cb326-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">colour =</span> occurrences, <span class="at">x =</span> df_sites<span class="sc">$</span>lon,</span>
<span id="cb326-5"><a href="ch-03.html#cb326-5" aria-hidden="true" tabindex="-1"></a>                              <span class="at">y =</span> df_sites<span class="sc">$</span>lat)) <span class="sc">+</span></span>
<span id="cb326-6"><a href="ch-03.html#cb326-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_gradient</span>(<span class="at">low =</span> <span class="st">&quot;yellow&quot;</span>, <span class="at">high =</span> <span class="st">&quot;red&quot;</span>, <span class="at">na.value =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb326-7"><a href="ch-03.html#cb326-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Longitude&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Latitude&quot;</span>) <span class="sc">+</span> <span class="fu">coord_quickmap</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-204-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>We see that the numbers of occurrence records in southern Europe are lower than in central Europe. This suggests there is a sampling bias, with a higher sampling effort in central European countries.</p>
<p>Knowing the species richness of the different sites can help us learn more about the ecological context the towers are located in. We made the occurrence matrix above and now we’ll use that to plot the richness. The richness is the number of species at each site. Since each row in the matrix is a site we can just sum the rows and we’ll get the number of species per site. We don’t have data for all our towers in <em>df_sites</em> so we will make a new dataframe with only the relevant sites and call it <em>df_sp_sites</em>.</p>
<div class="sourceCode" id="cb327"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb327-1"><a href="ch-03.html#cb327-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the sites for which we have no species lists</span></span>
<span id="cb327-2"><a href="ch-03.html#cb327-2" aria-hidden="true" tabindex="-1"></a>df_sp_sites <span class="ot">&lt;-</span> df_sites <span class="sc">%&gt;%</span> </span>
<span id="cb327-3"><a href="ch-03.html#cb327-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span>(site <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;CH-Oe1&quot;</span>, <span class="st">&quot;DE-Obe&quot;</span>, <span class="st">&quot;FR-LBr&quot;</span>, <span class="st">&quot;IT-Cpz&quot;</span>, <span class="st">&quot;IT-Ro1&quot;</span>)))</span>
<span id="cb327-4"><a href="ch-03.html#cb327-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb327-5"><a href="ch-03.html#cb327-5" aria-hidden="true" tabindex="-1"></a><span class="co"># the first column is the &#39;id&#39;, so we exclude that</span></span>
<span id="cb327-6"><a href="ch-03.html#cb327-6" aria-hidden="true" tabindex="-1"></a>richness <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(species_occ_mat[,<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb327-7"><a href="ch-03.html#cb327-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb327-8"><a href="ch-03.html#cb327-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb327-9"><a href="ch-03.html#cb327-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_polygon</span>(<span class="at">data =</span> europe_shape,</span>
<span id="cb327-10"><a href="ch-03.html#cb327-10" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x =</span> long, <span class="at">y =</span> lat, <span class="at">group =</span> group), <span class="at">fill =</span> <span class="cn">NA</span>, <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb327-11"><a href="ch-03.html#cb327-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">colour =</span> richness, <span class="at">x =</span> df_sp_sites<span class="sc">$</span>lon,</span>
<span id="cb327-12"><a href="ch-03.html#cb327-12" aria-hidden="true" tabindex="-1"></a>                              <span class="at">y =</span> df_sp_sites<span class="sc">$</span>lat)) <span class="sc">+</span></span>
<span id="cb327-13"><a href="ch-03.html#cb327-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_gradient</span>(<span class="at">low =</span> <span class="st">&quot;yellow&quot;</span>, <span class="at">high =</span> <span class="st">&quot;red&quot;</span>, <span class="at">na.value =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb327-14"><a href="ch-03.html#cb327-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Longitude&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Latitude&quot;</span>) <span class="sc">+</span> <span class="fu">coord_quickmap</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-205-1.png" width="50%" style="display: block; margin: auto;" /></p>
</div>
<div id="species-traits" class="section level4" number="3.2.7.1">
<h4><span class="header-section-number">3.2.7.1</span> Species Traits</h4>
<p>In the next step, we will study the relationship between plant traits, GPP, and climatic variables for our fluxnet towers. We obtained data from the TRY database, which is a database containing plant trait data. You can make data requests on <a href="www.https://www.try-db.org">their website</a>. You can select the traits that you are interested in and the species for which you would like to look at the traits.
For this tutorial we already requested a dataset, because usually, it takes time to receive the data. We downloaded data on 9 different traits for all species available. We extracted the trait data for all the species we have in our species list (and therefore also in our species occurrence matrix).</p>
<p>Let’s load the data.</p>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb328-1"><a href="ch-03.html#cb328-1" aria-hidden="true" tabindex="-1"></a>trait_df <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;./data/Trait_df.csv&quot;</span>)</span>
<span id="cb328-2"><a href="ch-03.html#cb328-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(trait_df) <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">5</span>) <span class="co"># Only displaying 4 of 9 features</span></span></code></pre></div>
<pre><code>##   X.1 X         SpeciesName TraitID
## 1   1 1      Acer campestre      46
## 2   2 2      Acer campestre       2
## 3   3 3 Acer pseudoplatanus      46
## 4   4 4 Acer pseudoplatanus       2
## 5   5 5     Alnus glutinosa      46
## 6   6 6     Alnus glutinosa       2</code></pre>
<p>Some examples of traits are are leaf texture, leaf thickness, eaf chlorophyll content per leaf area, specific leaf area, etc.</p>
<p>As a next step, we can combine the species occurrence data with the trait data. The presence/absence matrix (<em>species_occ_mat</em>) we made earlier, gives us the species present at each site. From that we can then extract these species and their respective trait values from the trait dataframe (<em>trait_df</em>). Then we go on to summarise this data by calculating the mean value for each trait across all the species at the tower sites.</p>
<p>We start by doing this for one site.</p>
<div class="sourceCode" id="cb330"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb330-1"><a href="ch-03.html#cb330-1" aria-hidden="true" tabindex="-1"></a>ba <span class="ot">&lt;-</span> trait_df</span>
<span id="cb330-2"><a href="ch-03.html#cb330-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb330-3"><a href="ch-03.html#cb330-3" aria-hidden="true" tabindex="-1"></a><span class="co"># select all the species (species names= column names) which are present (so ==1)</span></span>
<span id="cb330-4"><a href="ch-03.html#cb330-4" aria-hidden="true" tabindex="-1"></a>species_present <span class="ot">&lt;-</span> <span class="fu">colnames</span>(species_occ_mat)[<span class="fu">which</span>(species_occ_mat[<span class="dv">1</span>, ] <span class="sc">==</span> <span class="dv">1</span>)]</span>
<span id="cb330-5"><a href="ch-03.html#cb330-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb330-6"><a href="ch-03.html#cb330-6" aria-hidden="true" tabindex="-1"></a><span class="co"># filter out the trait data for the species present at that site</span></span>
<span id="cb330-7"><a href="ch-03.html#cb330-7" aria-hidden="true" tabindex="-1"></a>trait_subset <span class="ot">&lt;-</span> ba[ba<span class="sc">$</span>SpeciesName <span class="sc">%in%</span> species_present,]</span>
<span id="cb330-8"><a href="ch-03.html#cb330-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb330-9"><a href="ch-03.html#cb330-9" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the mean of each trait across all species present</span></span>
<span id="cb330-10"><a href="ch-03.html#cb330-10" aria-hidden="true" tabindex="-1"></a>bac <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(trait_subset<span class="sc">$</span>OrigValueStr, <span class="fu">list</span>(trait_subset<span class="sc">$</span>TraitID), mean, <span class="at">na.rm=</span>T)</span></code></pre></div>
<p>To finish preparing the data for the PCA, we do the above steps for all the remaining sites by using a <em>‘for loop’</em>.</p>
<div class="sourceCode" id="cb331"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb331-1"><a href="ch-03.html#cb331-1" aria-hidden="true" tabindex="-1"></a>df_pca <span class="ot">&lt;-</span> bac<span class="sc">$</span>x</span>
<span id="cb331-2"><a href="ch-03.html#cb331-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">28</span>)){</span>
<span id="cb331-3"><a href="ch-03.html#cb331-3" aria-hidden="true" tabindex="-1"></a>    species_present <span class="ot">&lt;-</span> <span class="fu">colnames</span>(species_occ_mat)[<span class="fu">which</span>(species_occ_mat[i, ] <span class="sc">==</span> <span class="dv">1</span>)]</span>
<span id="cb331-4"><a href="ch-03.html#cb331-4" aria-hidden="true" tabindex="-1"></a>    trait_subset <span class="ot">&lt;-</span> ba[ba<span class="sc">$</span>SpeciesName <span class="sc">%in%</span> species_present,]</span>
<span id="cb331-5"><a href="ch-03.html#cb331-5" aria-hidden="true" tabindex="-1"></a>    bac <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(trait_subset<span class="sc">$</span>OrigValueStr, <span class="fu">list</span>(trait_subset<span class="sc">$</span>TraitID), mean, <span class="at">na.rm=</span>T)</span>
<span id="cb331-6"><a href="ch-03.html#cb331-6" aria-hidden="true" tabindex="-1"></a>    df_pca <span class="ot">&lt;-</span> <span class="fu">rbind</span>(df_pca, bac<span class="sc">$</span>x)</span>
<span id="cb331-7"><a href="ch-03.html#cb331-7" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>For the PCA we use only the traits with traitIDs: 46, 47, 3106 and 3117. This is because the other traits are just slightly modified versions of these four traits. These four traitIDs correspond to the traits: <em>leaf thickness</em>, <em>leaf dry matter content (LDMC)</em>, <em>plant height</em> and <em>specific leaf area (SLA)</em>. <em>Leaf dry matter content</em> is calculated as leaf dry mass per leaf fresh mass and <em>specific leaf area</em> is defined as leaf area per leaf dry mass.</p>
<div class="sourceCode" id="cb332"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb332-1"><a href="ch-03.html#cb332-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we select the columns corresponding to our desired traits</span></span>
<span id="cb332-2"><a href="ch-03.html#cb332-2" aria-hidden="true" tabindex="-1"></a>df_pca <span class="ot">&lt;-</span> <span class="fu">cbind</span>(df_pca[,<span class="dv">2</span>],df_pca[,<span class="dv">3</span>],df_pca[,<span class="dv">5</span>],df_pca[,<span class="dv">9</span>])</span>
<span id="cb332-3"><a href="ch-03.html#cb332-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb332-4"><a href="ch-03.html#cb332-4" aria-hidden="true" tabindex="-1"></a><span class="co"># to make the data easier to understand we name the columns after the trait</span></span>
<span id="cb332-5"><a href="ch-03.html#cb332-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(df_pca) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;leaf thickness&quot;</span>, <span class="st">&quot;leaf dry matter content (LDMC)&quot;</span>, <span class="st">&quot;plant height vegetative&quot;</span>, <span class="st">&quot;specific leaf area (SLA)&quot;</span>)</span>
<span id="cb332-6"><a href="ch-03.html#cb332-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb332-7"><a href="ch-03.html#cb332-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_pca)</span></code></pre></div>
<pre><code>##        leaf thickness leaf dry matter content (LDMC) plant height vegetative
## df_pca       18.38967                       142.1853                80.59372
##              19.11832                       144.8715                80.23070
##              19.32090                       176.7112                80.68073
##              21.47128                       144.4585                73.76227
##              22.30091                       148.8697                70.61044
##              22.17279                       144.7945                73.61253
##        specific leaf area (SLA)
## df_pca                 77.28056
##                        78.31965
##                        81.82512
##                        78.28065
##                        76.57826
##                        76.61992</code></pre>
<p>Principal Component Analysis (PCA) is one of famous techniques for dimension reduction, feature extraction, and data visualization. In general, PCA is defined as the transformation of a high dimensional vector space to a low dimensional space. Just imagine having to visualise data with 10 dimensions… we often stuggle already with three. It’s almost impossible to effectively show the shape of such a high dimensional data distribution. PCA provides an efficient way to reduce the dimensionalty (i.e. from 10 to 2), so it is much easier to visualize the shape of data distribution. In a nutshell, this reduction happens by the PCA taking the two axes that best explain the variation in the data.</p>
<p>PCA is also useful in the modeling a robust classifier where considerably small number of high dimensional training data is provided. By reducing the dimensions of learning data sets, PCA provides an effective and efficient method for data description and classification.
For a more mathematical description of PCA you can check <a href="https://www.projectrhea.org/rhea/index.php/PCA_Theory_Examples" class="uri">https://www.projectrhea.org/rhea/index.php/PCA_Theory_Examples</a>.</p>
<p>So now let’s reduce the dimensions of our trait data as an example… For this we use the function <code>prcomp()</code>.</p>
<div class="sourceCode" id="cb334"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb334-1"><a href="ch-03.html#cb334-1" aria-hidden="true" tabindex="-1"></a>traits_pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(df_pca, <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="cn">TRUE</span>)</span>
<span id="cb334-2"><a href="ch-03.html#cb334-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(traits_pca)</span></code></pre></div>
<pre><code>## Importance of components:
##                           PC1    PC2    PC3     PC4
## Standard deviation     1.5372 0.9280 0.6972 0.53830
## Proportion of Variance 0.5907 0.2153 0.1215 0.07244
## Cumulative Proportion  0.5907 0.8060 0.9276 1.00000</code></pre>
<p>From the information above we can see how much of the variation in the data is explained by each axis or dimensionality. Ideally, you want the first to axes to explain the vast majority of the variation, while the subsequent ones are much less important.
But what do all these numbers mean and how does the data look. It’s time to visualise the PCA.</p>
<div class="sourceCode" id="cb336"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb336-1"><a href="ch-03.html#cb336-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggfortify)</span>
<span id="cb336-2"><a href="ch-03.html#cb336-2" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(traits_pca, <span class="at">data =</span> df_pca, <span class="at">col =</span><span class="st">&quot;#138D37&quot;</span>, <span class="at">loadings =</span> <span class="cn">TRUE</span>, <span class="at">loadings.colour =</span> <span class="st">&#39;blue&#39;</span>, <span class="at">loadings.label =</span> <span class="cn">TRUE</span>, <span class="at">loadings.label.size =</span> <span class="dv">3</span>) <span class="sc">+</span> <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-211-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>We have successfully reduced the dimensionality of our data! We started with four traits and now we have reduced the variation to two main axes. We see that PC1 is mainly associated to leaf dry matter content, and plant height, while PC2 is associated with specific leaf area and leaf thickness.</p>
<p>It can be interesting to see how the points relate spatially, this could, for example, allow us to identify clusters. Therefore, we will visualise our tower sites coloured by the corresponding position (positive or negative) on the PC1 axes.</p>
<div class="sourceCode" id="cb337"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb337-1"><a href="ch-03.html#cb337-1" aria-hidden="true" tabindex="-1"></a><span class="co"># start by designating trait and axis values to vector objects</span></span>
<span id="cb337-2"><a href="ch-03.html#cb337-2" aria-hidden="true" tabindex="-1"></a>tr1 <span class="ot">&lt;-</span> df_pca[,<span class="dv">1</span>] <span class="co"># these are the traits</span></span>
<span id="cb337-3"><a href="ch-03.html#cb337-3" aria-hidden="true" tabindex="-1"></a>tr2 <span class="ot">&lt;-</span> df_pca[,<span class="dv">2</span>] <span class="co"># these are the traits</span></span>
<span id="cb337-4"><a href="ch-03.html#cb337-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb337-5"><a href="ch-03.html#cb337-5" aria-hidden="true" tabindex="-1"></a>pc1 <span class="ot">&lt;-</span> traits_pca<span class="sc">$</span>x[,<span class="dv">1</span>] <span class="co"># these are the axis values</span></span>
<span id="cb337-6"><a href="ch-03.html#cb337-6" aria-hidden="true" tabindex="-1"></a>pc2 <span class="ot">&lt;-</span> traits_pca<span class="sc">$</span>x[,<span class="dv">2</span>] <span class="co"># these are the axis values</span></span></code></pre></div>
<div class="sourceCode" id="cb338"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb338-1"><a href="ch-03.html#cb338-1" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the colour palettes for positive and negative values</span></span>
<span id="cb338-2"><a href="ch-03.html#cb338-2" aria-hidden="true" tabindex="-1"></a>pospal <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;#E9E15C&quot;</span>, <span class="st">&quot;#CF2B1E&quot;</span>))</span>
<span id="cb338-3"><a href="ch-03.html#cb338-3" aria-hidden="true" tabindex="-1"></a>negpal <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;#1E3498&quot;</span>, <span class="st">&quot;#95D7DB&quot;</span>))</span>
<span id="cb338-4"><a href="ch-03.html#cb338-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb338-5"><a href="ch-03.html#cb338-5" aria-hidden="true" tabindex="-1"></a><span class="co"># we separate positive from negative values for visualization</span></span>
<span id="cb338-6"><a href="ch-03.html#cb338-6" aria-hidden="true" tabindex="-1"></a>neg_pc1 <span class="ot">&lt;-</span> pc1 </span>
<span id="cb338-7"><a href="ch-03.html#cb338-7" aria-hidden="true" tabindex="-1"></a>neg_pc1[neg_pc1 <span class="sc">&gt;</span> <span class="dv">0</span> ] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb338-8"><a href="ch-03.html#cb338-8" aria-hidden="true" tabindex="-1"></a>pos_pc1 <span class="ot">&lt;-</span> pc1 </span>
<span id="cb338-9"><a href="ch-03.html#cb338-9" aria-hidden="true" tabindex="-1"></a>pos_pc1[pos_pc1 <span class="sc">&lt;</span> <span class="dv">0</span> ] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb338-10"><a href="ch-03.html#cb338-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb338-11"><a href="ch-03.html#cb338-11" aria-hidden="true" tabindex="-1"></a>datColpos <span class="ot">&lt;-</span> <span class="fu">pospal</span>(<span class="dv">100</span>)[<span class="fu">as.numeric</span>(<span class="fu">cut</span>(pos_pc1,<span class="at">breaks =</span> <span class="dv">100</span>))] </span>
<span id="cb338-12"><a href="ch-03.html#cb338-12" aria-hidden="true" tabindex="-1"></a>datColneg <span class="ot">&lt;-</span> <span class="fu">negpal</span>(<span class="dv">100</span>)[<span class="fu">as.numeric</span>(<span class="fu">cut</span>(neg_pc1,<span class="at">breaks =</span> <span class="dv">100</span>))]</span></code></pre></div>
<p>Once the colours are defined, we can plot PC1.</p>
<div class="sourceCode" id="cb339"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb339-1"><a href="ch-03.html#cb339-1" aria-hidden="true" tabindex="-1"></a><span class="co">#PC1 </span></span>
<span id="cb339-2"><a href="ch-03.html#cb339-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(europe_shape)</span>
<span id="cb339-3"><a href="ch-03.html#cb339-3" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(df_sp_sites<span class="sc">$</span>lon, df_sp_sites<span class="sc">$</span>lat, <span class="at">col=</span>datColpos,<span class="at">cex=</span><span class="dv">1</span>, <span class="at">pch =</span> <span class="dv">19</span>) </span>
<span id="cb339-4"><a href="ch-03.html#cb339-4" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(df_sp_sites<span class="sc">$</span>lon, df_sp_sites<span class="sc">$</span>lat, <span class="at">col=</span>datColneg,<span class="at">cex=</span><span class="dv">1</span>, <span class="at">pch =</span> <span class="dv">19</span>)</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-214-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>And then repeat the process for PC2:</p>
<div class="sourceCode" id="cb340"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb340-1"><a href="ch-03.html#cb340-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PC2</span></span>
<span id="cb340-2"><a href="ch-03.html#cb340-2" aria-hidden="true" tabindex="-1"></a>neg_pc2 <span class="ot">&lt;-</span> pc2 </span>
<span id="cb340-3"><a href="ch-03.html#cb340-3" aria-hidden="true" tabindex="-1"></a>neg_pc2[neg_pc2 <span class="sc">&gt;</span> <span class="dv">0</span> ] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb340-4"><a href="ch-03.html#cb340-4" aria-hidden="true" tabindex="-1"></a>pos_pc2 <span class="ot">&lt;-</span> pc2 </span>
<span id="cb340-5"><a href="ch-03.html#cb340-5" aria-hidden="true" tabindex="-1"></a>pos_pc2[pos_pc2 <span class="sc">&lt;</span> <span class="dv">0</span> ] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb340-6"><a href="ch-03.html#cb340-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb340-7"><a href="ch-03.html#cb340-7" aria-hidden="true" tabindex="-1"></a>datColpos <span class="ot">&lt;-</span> <span class="fu">pospal</span>(<span class="dv">100</span>)[<span class="fu">as.numeric</span>(<span class="fu">cut</span>(pos_pc2,<span class="at">breaks =</span> <span class="dv">100</span>))] </span>
<span id="cb340-8"><a href="ch-03.html#cb340-8" aria-hidden="true" tabindex="-1"></a>datColneg <span class="ot">&lt;-</span> <span class="fu">negpal</span>(<span class="dv">100</span>)[<span class="fu">as.numeric</span>(<span class="fu">cut</span>(neg_pc2,<span class="at">breaks =</span> <span class="dv">100</span>))]</span>
<span id="cb340-9"><a href="ch-03.html#cb340-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb340-10"><a href="ch-03.html#cb340-10" aria-hidden="true" tabindex="-1"></a><span class="co"># and plot it on the map</span></span>
<span id="cb340-11"><a href="ch-03.html#cb340-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(europe_shape)</span>
<span id="cb340-12"><a href="ch-03.html#cb340-12" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(df_sp_sites<span class="sc">$</span>lon, df_sp_sites<span class="sc">$</span>lat, <span class="at">col=</span>datColpos,<span class="at">cex=</span><span class="dv">1</span>, <span class="at">pch =</span> <span class="dv">19</span>)</span>
<span id="cb340-13"><a href="ch-03.html#cb340-13" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(df_sp_sites<span class="sc">$</span>lon, df_sp_sites<span class="sc">$</span>lat, <span class="at">col=</span>datColneg,<span class="at">cex=</span><span class="dv">1</span>, <span class="at">pch =</span> <span class="dv">19</span>)</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-215-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>The last step in our trait data PCA is to compare the results with GPP. Since GPP has more sites than we used for this analysis we start by removing these sites.</p>
<div class="sourceCode" id="cb341"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb341-1"><a href="ch-03.html#cb341-1" aria-hidden="true" tabindex="-1"></a>GPP_pca <span class="ot">&lt;-</span> GPP <span class="sc">%&gt;%</span> </span>
<span id="cb341-2"><a href="ch-03.html#cb341-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span>(siteid <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;CH-Oe1&quot;</span>, <span class="st">&quot;DE-Obe&quot;</span>, <span class="st">&quot;FR-LBr&quot;</span>, <span class="st">&quot;IT-Cpz&quot;</span>, <span class="st">&quot;IT-Ro1&quot;</span>)))</span></code></pre></div>
<p>Then we make a linear model for PC1 and GPP and plot it.</p>
<div class="sourceCode" id="cb342"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb342-1"><a href="ch-03.html#cb342-1" aria-hidden="true" tabindex="-1"></a>reg_gpp_pc1 <span class="ot">&lt;-</span> <span class="fu">lm</span>(gpp_meanann<span class="sc">~</span>pc1, <span class="at">data =</span> GPP_pca)</span>
<span id="cb342-2"><a href="ch-03.html#cb342-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-3"><a href="ch-03.html#cb342-3" aria-hidden="true" tabindex="-1"></a>GPP_pca <span class="sc">%&gt;%</span> </span>
<span id="cb342-4"><a href="ch-03.html#cb342-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> pc1, <span class="at">y =</span> gpp_meanann)) <span class="sc">+</span></span>
<span id="cb342-5"><a href="ch-03.html#cb342-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb342-6"><a href="ch-03.html#cb342-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span><span class="st">&#39;lm&#39;</span>, <span class="at">color=</span><span class="st">&quot;red&quot;</span>, <span class="at">size=</span><span class="fl">0.5</span>, <span class="at">se=</span><span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb342-7"><a href="ch-03.html#cb342-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;PC1&quot;</span>) <span class="sc">+</span></span>
<span id="cb342-8"><a href="ch-03.html#cb342-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;GPP mean [µmolCO2 m-2 s-1]&quot;</span>) <span class="sc">+</span></span>
<span id="cb342-9"><a href="ch-03.html#cb342-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-217-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>We repeat this for GPP and PC2.</p>
<div class="sourceCode" id="cb343"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb343-1"><a href="ch-03.html#cb343-1" aria-hidden="true" tabindex="-1"></a>reg_gpp_pc2 <span class="ot">&lt;-</span> <span class="fu">lm</span>(gpp_meanann<span class="sc">~</span>pc2, <span class="at">data =</span> GPP_pca)</span>
<span id="cb343-2"><a href="ch-03.html#cb343-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb343-3"><a href="ch-03.html#cb343-3" aria-hidden="true" tabindex="-1"></a>GPP_pca <span class="sc">%&gt;%</span> </span>
<span id="cb343-4"><a href="ch-03.html#cb343-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> pc2, <span class="at">y =</span> gpp_meanann)) <span class="sc">+</span></span>
<span id="cb343-5"><a href="ch-03.html#cb343-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb343-6"><a href="ch-03.html#cb343-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span><span class="st">&#39;lm&#39;</span>, <span class="at">color=</span><span class="st">&quot;red&quot;</span>, <span class="at">size=</span><span class="fl">0.5</span>, <span class="at">se=</span><span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb343-7"><a href="ch-03.html#cb343-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;PC2&quot;</span>) <span class="sc">+</span></span>
<span id="cb343-8"><a href="ch-03.html#cb343-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;GPP mean [µmolCO2 m-2 s-1]&quot;</span>) <span class="sc">+</span></span>
<span id="cb343-9"><a href="ch-03.html#cb343-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() </span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-218-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>Once again we see weak correlations with GPP. Therefore, we can conclude that this further search for predictors that explain the variation in GPP has not yielded any good results. We will have to continue looking for better predictors, which explain this variation in GPP between the sites.</p>
<p><strong>Checkpoint</strong></p>
<p>Plot the spatial distribution of the trait ‘leaf thickness,’ make the colour or size related to the leaf thickness values. Describe what you can gather from this plot.</p>
<p><strong>Solution</strong></p>
<div class="sourceCode" id="cb344"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb344-1"><a href="ch-03.html#cb344-1" aria-hidden="true" tabindex="-1"></a>leaf_t <span class="ot">&lt;-</span> df_pca[,<span class="dv">1</span>]</span>
<span id="cb344-2"><a href="ch-03.html#cb344-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb344-3"><a href="ch-03.html#cb344-3" aria-hidden="true" tabindex="-1"></a>leavepal <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;#E2C151&quot;</span>,<span class="st">&quot;#138D37&quot;</span>))</span>
<span id="cb344-4"><a href="ch-03.html#cb344-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb344-5"><a href="ch-03.html#cb344-5" aria-hidden="true" tabindex="-1"></a>datColleaves <span class="ot">&lt;-</span> <span class="fu">leavepal</span>(<span class="dv">10</span>)[<span class="fu">as.numeric</span>(<span class="fu">cut</span>(leaf_t,<span class="at">breaks =</span> <span class="dv">10</span>))]</span>
<span id="cb344-6"><a href="ch-03.html#cb344-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb344-7"><a href="ch-03.html#cb344-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(europe_shape)</span>
<span id="cb344-8"><a href="ch-03.html#cb344-8" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(df_sp_sites<span class="sc">$</span>lon, df_sp_sites<span class="sc">$</span>lat, <span class="at">col=</span>datColleaves,<span class="at">cex=</span>leaf_t<span class="sc">/</span><span class="dv">18</span>, <span class="at">pch =</span> <span class="dv">19</span>) </span></code></pre></div>
<p><img src="esds_book_files/figure-html/unnamed-chunk-219-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>The distribution of leaf thickness seems to depend on latitude. In Southern regions, leaves are generally thinner than in Northern regions. To confirm this we would have to make a model and test our observation.</p>
</div>
</div>
</div>
<div id="exercise-2" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Exercise</h2>
<div id="part-1-plotting-elevation-differences" class="section level3" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Part 1: Plotting Elevation differences</h3>
<p>In this first part of the exercise, we consider elevation data from the ETOPO dataset. ETOPO is a 1 arc-minute global relief model of the Earth’s surface that integrates land topography and ocean bathymetry. Since it is modelled date there is a certain level of uncertainty associated with it. In this exercise, you will look at the differences between the elevations from the tower data, which is directly measured at the site, and the values you will extract from this modelled ETOPO data.</p>
<p><em>Note:</em> think about which packages you need to load to complete each of the tasks.</p>
<ol style="list-style-type: lower-alpha">
<li><p>Load the <em>‘dataset_ex3.csv’</em> file containing the elevations of the towers.</p></li>
<li><p>Load both the <em>‘ETOPO.tif’</em> raster and the <em>‘europe_shape’</em> shapefile. Plot the shape of Europe over the raster.
(Since we used the shapefile in the tutorial too, it isn’t located in the same folder as the data for this exercise. You’ll have to use the correct pathway to load it, if you are unsure about how to find this use the ‘Files’ tab on the right can help you figure it out.)</p></li>
<li><p>Extract the elevations from the raster at the tower sites.</p></li>
<li><p>Calculate the difference between the elevations of the towers and the one extracted from the ETOPO.</p></li>
<li><p>Plot the difference in elevation calculated for each site on a map of Europe.
<em>Hint:</em> there are several ways to do this, one could be to using ifelse().</p></li>
</ol>
</div>
<div id="part-2-temperature-and-elevation-correlations" class="section level3" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> Part 2: Temperature and Elevation Correlations</h3>
<p>In this second part, we want to find out which of the elevations we gathered in the first part (ETOPO and towers) correlates better with the temperatures measured at the towers.</p>
<ol style="list-style-type: lower-alpha">
<li><p>Correlate the ETOPO elevations at the towers sites with the temperatures measured at the towers.</p></li>
<li><p>Correlate the elevations measured at the tower sites with the temperatures measured at the towers.</p></li>
<li><p>What is the mean squared errors of the two models above. Which of the two correlates better?</p></li>
<li><p>The <em>‘lapse rate’</em> is known as the rate of decrease of an atmospheric variable, in our case temperature, with increased elevation (also altitude). This <em>lapse</em> (gradual fall) in temperature with increasing altitude at a given time and place, is the slope of your models above. Find and compare the <em>lapse rate</em> of the two models. <em>Hint:</em> if you are stuck take a closer look at your model with summary().</p></li>
</ol>
<p>Now, that you’ve made it to the end of the exercise, remember to ‘knit’ this document so you can upload the html to moodle.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="ch-02.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="ch-04.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
